{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d182c426-4d95-4dee-a099-cd923e9437d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Script for Pathway enrichment analysis on MOFA results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "213d69de-4705-4a6e-9387-b1a3413172eb",
   "metadata": {
    "tags": []
   },
   "source": [
    "#############################################\n",
    "# Prerequisites - Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "49f265d4-5df8-446f-ad65-4a49276593b0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on '../Global_Configs.csv'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/jove_r//lib/R/library\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.4; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ───────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ─────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m      masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m       masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mhour()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::hour()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32misoweek()\u001b[39m masks \u001b[34mdata.table\u001b[39m::isoweek()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m         masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m        masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmday()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::mday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mminute()\u001b[39m  masks \u001b[34mdata.table\u001b[39m::minute()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mmonth()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::month()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mquarter()\u001b[39m masks \u001b[34mdata.table\u001b[39m::quarter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m  masks \u001b[34mdata.table\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mwday()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::wday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mweek()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::week()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myday()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::yday()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32myear()\u001b[39m    masks \u001b[34mdata.table\u001b[39m::year()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "\n",
      "Attaching package: ‘caret’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    lift\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    discard\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:readr’:\n",
      "\n",
      "    col_factor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    geometry\n",
      "\n",
      "\n",
      "corrplot 0.92 loaded\n",
      "\n",
      "\n",
      "Attaching package: ‘MOFA2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    predict\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Warning message in checkDepPackageVersion(dep_pkg = \"TMB\"):\n",
      "“Package version inconsistency detected.\n",
      "glmmTMB was built with TMB version 1.9.6\n",
      "Current TMB version is 1.9.10\n",
      "Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)”\n",
      "Registered S3 method overwritten by 'EnvStats':\n",
      "  method         from\n",
      "  print.estimate lava\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab19cd2b-a715-46d7-ab96-834040cc5018",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `linewidth` argument instead.”\n"
     ]
    }
   ],
   "source": [
    "source('MS2_Plot_Config.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f0a37ca6-9dd3-48e3-b4f8-e13ce1f4a24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d409d14-a693-4722-9648-4b6bacb4b74a",
   "metadata": {
    "tags": []
   },
   "source": [
    "###############################################\n",
    "# Preqrequisites Configurations & Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "578861cf-62ac-49a2-aaa6-0c91e751416b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on 'configurations/Global_Configs.csv'”\n"
     ]
    }
   ],
   "source": [
    "### Load configurations file\n",
    "global_configs = read.csv('configurations/Global_Configs.csv', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1306e099-a83f-40ce-8ea6-7fe702c88747",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>parameter</th><th scope=col>parameter.1</th><th scope=col>value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>data_path  </td><td>data_path  </td><td>../results/data/input_data/</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>result_path</td><td>result_path</td><td>../results/data/results/   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & parameter & parameter.1 & value\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & data\\_path   & data\\_path   & ../results/data/input\\_data/\\\\\n",
       "\t2 & result\\_path & result\\_path & ../results/data/results/   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | parameter &lt;chr&gt; | parameter.1 &lt;chr&gt; | value &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | data_path   | data_path   | ../results/data/input_data/ |\n",
       "| 2 | result_path | result_path | ../results/data/results/    |\n",
       "\n"
      ],
      "text/plain": [
       "  parameter   parameter.1 value                      \n",
       "1 data_path   data_path   ../results/data/input_data/\n",
       "2 result_path result_path ../results/data/results/   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(global_configs,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1b570bb-26f4-4ba8-87ea-e4976278240b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = global_configs$value[global_configs$parameter == 'data_path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b212cfcc-0368-442f-ab1e-54127c0d8654",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../results/data/input_data/'"
      ],
      "text/latex": [
       "'../results/data/input\\_data/'"
      ],
      "text/markdown": [
       "'../results/data/input_data/'"
      ],
      "text/plain": [
       "[1] \"../results/data/input_data/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e96fa55c-83ba-464d-9a8b-2e4c584c519f",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path = global_configs$value[global_configs$parameter == 'result_path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7350f322-c30d-4bb9-8bc8-88b7df9df57e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../results/data/results/'"
      ],
      "text/latex": [
       "'../results/data/results/'"
      ],
      "text/markdown": [
       "'../results/data/results/'"
      ],
      "text/plain": [
       "[1] \"../results/data/results/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "result_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0929af17-aae4-4f1a-aa23-d9a7d4afa033",
   "metadata": {},
   "outputs": [],
   "source": [
    "###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "02e6bb78-83e9-47cf-8288-5c93e30ea7aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on 'configurations/06_Pathway_Configs.csv'”\n"
     ]
    }
   ],
   "source": [
    "pathway_configs = read.csv('configurations/06_Pathway_Configs.csv', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f9f4ee34-34a3-4a18-8954-47a920fd6e98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>mofa_result_name</th><th scope=col>factor_set</th><th scope=col>coverage_par</th><th scope=col>types</th><th scope=col>coverage_plot</th><th scope=col>p_value_plot</th><th scope=col>max_pathways_plot</th><th scope=col>enrichment_plot</th><th scope=col>top_features_plot</th><th scope=col>pathway_selection</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>config1</td><td>1,2,3,4,5</td><td>0.2</td><td>B.cell,CD16.Mono,CD4.CTL,CD4.Naive,CD4.TCM,CD4.TEM,cDC2,gdT,neutrophil,NK,Treg</td><td>0.5</td><td>0.05</td><td>8</td><td>positive</td><td>0.125</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & mofa\\_result\\_name & factor\\_set & coverage\\_par & types & coverage\\_plot & p\\_value\\_plot & max\\_pathways\\_plot & enrichment\\_plot & top\\_features\\_plot & pathway\\_selection\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <int> & <chr> & <dbl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & config1 & 1,2,3,4,5 & 0.2 & B.cell,CD16.Mono,CD4.CTL,CD4.Naive,CD4.TCM,CD4.TEM,cDC2,gdT,neutrophil,NK,Treg & 0.5 & 0.05 & 8 & positive & 0.125 & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 10\n",
       "\n",
       "| <!--/--> | mofa_result_name &lt;chr&gt; | factor_set &lt;chr&gt; | coverage_par &lt;dbl&gt; | types &lt;chr&gt; | coverage_plot &lt;dbl&gt; | p_value_plot &lt;dbl&gt; | max_pathways_plot &lt;int&gt; | enrichment_plot &lt;chr&gt; | top_features_plot &lt;dbl&gt; | pathway_selection &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | config1 | 1,2,3,4,5 | 0.2 | B.cell,CD16.Mono,CD4.CTL,CD4.Naive,CD4.TCM,CD4.TEM,cDC2,gdT,neutrophil,NK,Treg | 0.5 | 0.05 | 8 | positive | 0.125 | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  mofa_result_name factor_set coverage_par\n",
       "1 config1          1,2,3,4,5  0.2         \n",
       "  types                                                                         \n",
       "1 B.cell,CD16.Mono,CD4.CTL,CD4.Naive,CD4.TCM,CD4.TEM,cDC2,gdT,neutrophil,NK,Treg\n",
       "  coverage_plot p_value_plot max_pathways_plot enrichment_plot\n",
       "1 0.5           0.05         8                 positive       \n",
       "  top_features_plot pathway_selection\n",
       "1 0.125             NA               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(pathway_configs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47a498a2-73ff-4bed-9a9b-cd911f61db70",
   "metadata": {},
   "source": [
    "# Define parameters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cee901ad-739e-4409-9b63-72008da8d695",
   "metadata": {},
   "outputs": [],
   "source": [
    "## From Excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "864b5b1e-4f82-4f00-8481-7639a9ba85fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "df558a43-5a61-44a0-baae-bf28e9ec100c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mofa_name = pathway_configs$mofa_result_name[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4e086640-cbd1-4541-b5da-99b3f355a328",
   "metadata": {},
   "outputs": [],
   "source": [
    "factor_set = as.numeric(unlist(str_split(pathway_configs$factor_set[1], ',')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a0e2a3f1-001a-4e4c-b8ed-cb86127b3571",
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_par = pathway_configs$coverage_par[1] # less than 20 % "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4f1ca639-8ab9-49dc-82df-1d7a0c5f68ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "views_set = unlist(str_split( pathway_configs$types[1], ','))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cfade333-33ee-4f75-b4c8-71e11c2a37fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B.cell'</li><li>'CD16.Mono'</li><li>'CD4.CTL'</li><li>'CD4.Naive'</li><li>'CD4.TCM'</li><li>'CD4.TEM'</li><li>'cDC2'</li><li>'gdT'</li><li>'neutrophil'</li><li>'NK'</li><li>'Treg'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B.cell'\n",
       "\\item 'CD16.Mono'\n",
       "\\item 'CD4.CTL'\n",
       "\\item 'CD4.Naive'\n",
       "\\item 'CD4.TCM'\n",
       "\\item 'CD4.TEM'\n",
       "\\item 'cDC2'\n",
       "\\item 'gdT'\n",
       "\\item 'neutrophil'\n",
       "\\item 'NK'\n",
       "\\item 'Treg'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B.cell'\n",
       "2. 'CD16.Mono'\n",
       "3. 'CD4.CTL'\n",
       "4. 'CD4.Naive'\n",
       "5. 'CD4.TCM'\n",
       "6. 'CD4.TEM'\n",
       "7. 'cDC2'\n",
       "8. 'gdT'\n",
       "9. 'neutrophil'\n",
       "10. 'NK'\n",
       "11. 'Treg'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"B.cell\"     \"CD16.Mono\"  \"CD4.CTL\"    \"CD4.Naive\"  \"CD4.TCM\"   \n",
       " [6] \"CD4.TEM\"    \"cDC2\"       \"gdT\"        \"neutrophil\" \"NK\"        \n",
       "[11] \"Treg\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "views_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d48078ce-671e-416b-b7c5-757f7372d3da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bd3201a4-3fea-487c-9e1b-a05ab8859c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for visualization\n",
    "## Automatic\n",
    "coverage_par = pathway_configs$coverage_plot[1]\n",
    "p_value_cutoff_plot =pathway_configs$p_value_plot[1]\n",
    "max_pathways =pathway_configs$max_pathways_plot[1]\n",
    "select_enrichment = pathway_configs$enrichment_plot[1]\n",
    "\n",
    "## selected pathways\n",
    "pathway_selection_var =pathway_configs$pathway_selection[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d9298328-bdd3-4991-bb2b-966a20ccde78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "&lt;NA&gt;"
      ],
      "text/latex": [
       "<NA>"
      ],
      "text/markdown": [
       "&lt;NA&gt;"
      ],
      "text/plain": [
       "[1] NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pathway_selection_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7af39745-7dbc-4dce-a760-ed2d72e3976f",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_var_thres =pathway_configs$top_features_plot[1] # choose threshold of top x % of features of MOFA factor to take into account"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62072a57-a0f3-41ce-8699-d9678c1acda4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4da9d560-4ec4-40e7-9813-95e386e16db4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "35f43524-b597-4952-a2d5-6756148d9379",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Currently fixed in manuscript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "979816d7-fed6-4aa4-ab1a-ff84fd5831ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "## For enrichment calculation\n",
    "use_statistic = \"rank.sum\" # alternatives: mean.diff, rank.sum\n",
    "use_test = 'parametric'  # alternatives: permutation, parametric, \"cor.adj.parametric\"\n",
    "p_val_cutoff = 0.05\n",
    "min_size = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe4b50c-9602-40ac-a886-a2058dd80c5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "43e2937f-4e6d-4e90-99f2-2f7c370f9639",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load Data "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c532318-cf55-47b9-ba16-31b10391ff6a",
   "metadata": {},
   "source": [
    "## Model Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "54bbc68f-377a-4058-b5da-57ae261a1b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f0aeb981-5cd3-4e0d-a9a4-574335dcebd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name =  paste0(\"03_MOFA_MODEL_\", mofa_name,'.hdf5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "21486773-c9a7-41bd-a933-f01ac207b731",
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = file.path( paste0(result_path, '/03_results/',  model_name) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f7d86d16-52a8-425b-9855-794259a4b4ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../results/data/results//03_results/03_MOFA_MODEL_config1.hdf5'"
      ],
      "text/latex": [
       "'../results/data/results//03\\_results/03\\_MOFA\\_MODEL\\_config1.hdf5'"
      ],
      "text/markdown": [
       "'../results/data/results//03_results/03_MOFA_MODEL_config1.hdf5'"
      ],
      "text/plain": [
       "[1] \"../results/data/results//03_results/03_MOFA_MODEL_config1.hdf5\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8f2d4649-8f0c-416a-a211-c47fa63a7e41",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading data...\n",
      "\n",
      "Loading expectations for 2 nodes...\n",
      "\n",
      "Loading model options...\n",
      "\n",
      "Loading training options and statistics...\n",
      "\n",
      "Assigning names to the different dimensions...\n",
      "\n",
      "Re-ordering factors by their variance explained...\n",
      "\n",
      "Doing quality control...\n",
      "\n",
      "Checking views names...\n",
      "\n",
      "Checking groups names...\n",
      "\n",
      "Checking samples names...\n",
      "\n",
      "Checking features names...\n",
      "\n",
      "Checking dimensions...\n",
      "\n",
      "Checking there are no features with complete missing values...\n",
      "\n",
      "Checking sample covariates...\n",
      "\n",
      "Checking expectations...\n",
      "\n",
      "Checking for intercept factors...\n",
      "\n",
      "Warning message in .quality_control(object, verbose = verbose):\n",
      "“Factor(s) 5, 9, 10, 20 are strongly correlated with the total number of expressed features for at least one of your omics. Such factors appear when there are differences in the total 'levels' between your samples, *sometimes* because of poor normalisation in the preprocessing steps.\n",
      "”\n",
      "Checking for highly correlated factors...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model <- load_model(outfile, verbose = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70373296-fe48-46b9-9e5f-c0530518eff8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "35a71ce1-443b-4dda-8187-85c79c0e4f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways =  read.csv(paste0(data_path, 'Prepared_Pathway_Data.csv'))\n",
    "pathways$X = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "991b2cc4-e6cb-495c-8d52-75666f967096",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>gene</th><th scope=col>pathway_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-1059683</td><td>JAK1 </td><td>Interleukin-6 signaling</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>STAT1</td><td>Interleukin-6 signaling</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & ID & gene & pathway\\_name\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-1059683 & JAK1  & Interleukin-6 signaling\\\\\n",
       "\t2 & R-HSA-1059683 & STAT1 & Interleukin-6 signaling\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | gene &lt;chr&gt; | pathway_name &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | R-HSA-1059683 | JAK1  | Interleukin-6 signaling |\n",
       "| 2 | R-HSA-1059683 | STAT1 | Interleukin-6 signaling |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            gene  pathway_name           \n",
       "1 R-HSA-1059683 JAK1  Interleukin-6 signaling\n",
       "2 R-HSA-1059683 STAT1 Interleukin-6 signaling"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(pathways,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e3576c-af99-4c04-9b7b-35c06c7db34c",
   "metadata": {},
   "source": [
    "# Prepare model data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3f2a67b-a604-43cd-8231-a8fedd53e0ee",
   "metadata": {},
   "source": [
    "## Extract the weights from the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7fb063ef-e664-4c8d-bfd2-70287c7c3911",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 24</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Factor1</th><th scope=col>Factor2</th><th scope=col>Factor3</th><th scope=col>Factor4</th><th scope=col>Factor5</th><th scope=col>Factor6</th><th scope=col>Factor7</th><th scope=col>Factor8</th><th scope=col>Factor9</th><th scope=col>Factor10</th><th scope=col>⋯</th><th scope=col>Factor15</th><th scope=col>Factor16</th><th scope=col>Factor17</th><th scope=col>Factor18</th><th scope=col>Factor19</th><th scope=col>Factor20</th><th scope=col>type</th><th scope=col>variable_name</th><th scope=col>view</th><th scope=col>gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>B.cell__ACTB</th><td>-0.21732116</td><td> 0.18824680</td><td>-0.11174294</td><td> 0.080270097</td><td>-0.06805549</td><td>-0.0421615392</td><td>-0.064257897</td><td>-0.4432788</td><td> 0.007395637</td><td> 0.034436032</td><td>⋯</td><td> 0.01741151</td><td> 0.01803296</td><td> 0.07789122</td><td> 0.03405593</td><td>-0.14379902</td><td>-0.03069997</td><td>B.cell</td><td>B.cell__ACTB </td><td>B.cell</td><td>ACTB </td></tr>\n",
       "\t<tr><th scope=row>B.cell__ACTG1</th><td>-0.03822634</td><td>-0.02855239</td><td>-0.04507595</td><td> 0.062922463</td><td>-0.05201821</td><td>-0.0592270233</td><td> 0.067477025</td><td>-0.3285550</td><td>-0.012521170</td><td> 0.030991051</td><td>⋯</td><td>-0.02331705</td><td> 0.03813000</td><td> 0.10040610</td><td> 0.10898526</td><td> 0.02983564</td><td>-0.14744236</td><td>B.cell</td><td>B.cell__ACTG1</td><td>B.cell</td><td>ACTG1</td></tr>\n",
       "\t<tr><th scope=row>B.cell__ACTR2</th><td>-0.08317667</td><td> 0.21396732</td><td>-0.16865762</td><td> 0.004320409</td><td> 0.02414656</td><td> 0.0572508350</td><td> 0.007391086</td><td>-0.2363539</td><td> 0.001659897</td><td> 0.261096179</td><td>⋯</td><td> 0.02130334</td><td> 0.02949983</td><td>-0.06087974</td><td> 0.02812659</td><td>-0.13211621</td><td>-0.17683296</td><td>B.cell</td><td>B.cell__ACTR2</td><td>B.cell</td><td>ACTR2</td></tr>\n",
       "\t<tr><th scope=row>B.cell__ACTR3</th><td> 0.07242947</td><td> 0.04114458</td><td>-0.17622976</td><td>-0.093380645</td><td>-0.02262458</td><td> 0.0004040518</td><td> 0.008664441</td><td>-0.1160945</td><td> 0.010067453</td><td> 0.024094608</td><td>⋯</td><td>-0.08647003</td><td> 0.00610388</td><td>-0.05120527</td><td> 0.05020532</td><td> 0.02533577</td><td>-0.09245851</td><td>B.cell</td><td>B.cell__ACTR3</td><td>B.cell</td><td>ACTR3</td></tr>\n",
       "\t<tr><th scope=row>B.cell__ADD3</th><td>-0.10174651</td><td> 0.14137915</td><td>-0.18193237</td><td>-0.028330170</td><td>-0.03031231</td><td>-0.0162211861</td><td>-0.005273766</td><td>-0.2519083</td><td>-0.013892147</td><td>-0.003902171</td><td>⋯</td><td>-0.11196484</td><td>-0.02243120</td><td> 0.06199035</td><td>-0.02736166</td><td> 0.04897216</td><td>-0.04420387</td><td>B.cell</td><td>B.cell__ADD3 </td><td>B.cell</td><td>ADD3 </td></tr>\n",
       "\t<tr><th scope=row>B.cell__AES</th><td> 0.18231821</td><td>-0.12558725</td><td> 0.11752542</td><td> 0.090172626</td><td>-0.06976022</td><td> 0.0765330419</td><td> 0.073054507</td><td>-0.1399591</td><td>-0.022372559</td><td>-0.010901893</td><td>⋯</td><td> 0.12049376</td><td> 0.02960805</td><td> 0.04202096</td><td> 0.02239864</td><td>-0.05132128</td><td>-0.26910231</td><td>B.cell</td><td>B.cell__AES  </td><td>B.cell</td><td>AES  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 24\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & Factor1 & Factor2 & Factor3 & Factor4 & Factor5 & Factor6 & Factor7 & Factor8 & Factor9 & Factor10 & ⋯ & Factor15 & Factor16 & Factor17 & Factor18 & Factor19 & Factor20 & type & variable\\_name & view & gene\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tB.cell\\_\\_ACTB & -0.21732116 &  0.18824680 & -0.11174294 &  0.080270097 & -0.06805549 & -0.0421615392 & -0.064257897 & -0.4432788 &  0.007395637 &  0.034436032 & ⋯ &  0.01741151 &  0.01803296 &  0.07789122 &  0.03405593 & -0.14379902 & -0.03069997 & B.cell & B.cell\\_\\_ACTB  & B.cell & ACTB \\\\\n",
       "\tB.cell\\_\\_ACTG1 & -0.03822634 & -0.02855239 & -0.04507595 &  0.062922463 & -0.05201821 & -0.0592270233 &  0.067477025 & -0.3285550 & -0.012521170 &  0.030991051 & ⋯ & -0.02331705 &  0.03813000 &  0.10040610 &  0.10898526 &  0.02983564 & -0.14744236 & B.cell & B.cell\\_\\_ACTG1 & B.cell & ACTG1\\\\\n",
       "\tB.cell\\_\\_ACTR2 & -0.08317667 &  0.21396732 & -0.16865762 &  0.004320409 &  0.02414656 &  0.0572508350 &  0.007391086 & -0.2363539 &  0.001659897 &  0.261096179 & ⋯ &  0.02130334 &  0.02949983 & -0.06087974 &  0.02812659 & -0.13211621 & -0.17683296 & B.cell & B.cell\\_\\_ACTR2 & B.cell & ACTR2\\\\\n",
       "\tB.cell\\_\\_ACTR3 &  0.07242947 &  0.04114458 & -0.17622976 & -0.093380645 & -0.02262458 &  0.0004040518 &  0.008664441 & -0.1160945 &  0.010067453 &  0.024094608 & ⋯ & -0.08647003 &  0.00610388 & -0.05120527 &  0.05020532 &  0.02533577 & -0.09245851 & B.cell & B.cell\\_\\_ACTR3 & B.cell & ACTR3\\\\\n",
       "\tB.cell\\_\\_ADD3 & -0.10174651 &  0.14137915 & -0.18193237 & -0.028330170 & -0.03031231 & -0.0162211861 & -0.005273766 & -0.2519083 & -0.013892147 & -0.003902171 & ⋯ & -0.11196484 & -0.02243120 &  0.06199035 & -0.02736166 &  0.04897216 & -0.04420387 & B.cell & B.cell\\_\\_ADD3  & B.cell & ADD3 \\\\\n",
       "\tB.cell\\_\\_AES &  0.18231821 & -0.12558725 &  0.11752542 &  0.090172626 & -0.06976022 &  0.0765330419 &  0.073054507 & -0.1399591 & -0.022372559 & -0.010901893 & ⋯ &  0.12049376 &  0.02960805 &  0.04202096 &  0.02239864 & -0.05132128 & -0.26910231 & B.cell & B.cell\\_\\_AES   & B.cell & AES  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 24\n",
       "\n",
       "| <!--/--> | Factor1 &lt;dbl&gt; | Factor2 &lt;dbl&gt; | Factor3 &lt;dbl&gt; | Factor4 &lt;dbl&gt; | Factor5 &lt;dbl&gt; | Factor6 &lt;dbl&gt; | Factor7 &lt;dbl&gt; | Factor8 &lt;dbl&gt; | Factor9 &lt;dbl&gt; | Factor10 &lt;dbl&gt; | ⋯ ⋯ | Factor15 &lt;dbl&gt; | Factor16 &lt;dbl&gt; | Factor17 &lt;dbl&gt; | Factor18 &lt;dbl&gt; | Factor19 &lt;dbl&gt; | Factor20 &lt;dbl&gt; | type &lt;chr&gt; | variable_name &lt;chr&gt; | view &lt;chr&gt; | gene &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| B.cell__ACTB | -0.21732116 |  0.18824680 | -0.11174294 |  0.080270097 | -0.06805549 | -0.0421615392 | -0.064257897 | -0.4432788 |  0.007395637 |  0.034436032 | ⋯ |  0.01741151 |  0.01803296 |  0.07789122 |  0.03405593 | -0.14379902 | -0.03069997 | B.cell | B.cell__ACTB  | B.cell | ACTB  |\n",
       "| B.cell__ACTG1 | -0.03822634 | -0.02855239 | -0.04507595 |  0.062922463 | -0.05201821 | -0.0592270233 |  0.067477025 | -0.3285550 | -0.012521170 |  0.030991051 | ⋯ | -0.02331705 |  0.03813000 |  0.10040610 |  0.10898526 |  0.02983564 | -0.14744236 | B.cell | B.cell__ACTG1 | B.cell | ACTG1 |\n",
       "| B.cell__ACTR2 | -0.08317667 |  0.21396732 | -0.16865762 |  0.004320409 |  0.02414656 |  0.0572508350 |  0.007391086 | -0.2363539 |  0.001659897 |  0.261096179 | ⋯ |  0.02130334 |  0.02949983 | -0.06087974 |  0.02812659 | -0.13211621 | -0.17683296 | B.cell | B.cell__ACTR2 | B.cell | ACTR2 |\n",
       "| B.cell__ACTR3 |  0.07242947 |  0.04114458 | -0.17622976 | -0.093380645 | -0.02262458 |  0.0004040518 |  0.008664441 | -0.1160945 |  0.010067453 |  0.024094608 | ⋯ | -0.08647003 |  0.00610388 | -0.05120527 |  0.05020532 |  0.02533577 | -0.09245851 | B.cell | B.cell__ACTR3 | B.cell | ACTR3 |\n",
       "| B.cell__ADD3 | -0.10174651 |  0.14137915 | -0.18193237 | -0.028330170 | -0.03031231 | -0.0162211861 | -0.005273766 | -0.2519083 | -0.013892147 | -0.003902171 | ⋯ | -0.11196484 | -0.02243120 |  0.06199035 | -0.02736166 |  0.04897216 | -0.04420387 | B.cell | B.cell__ADD3  | B.cell | ADD3  |\n",
       "| B.cell__AES |  0.18231821 | -0.12558725 |  0.11752542 |  0.090172626 | -0.06976022 |  0.0765330419 |  0.073054507 | -0.1399591 | -0.022372559 | -0.010901893 | ⋯ |  0.12049376 |  0.02960805 |  0.04202096 |  0.02239864 | -0.05132128 | -0.26910231 | B.cell | B.cell__AES   | B.cell | AES   |\n",
       "\n"
      ],
      "text/plain": [
       "              Factor1     Factor2     Factor3     Factor4      Factor5    \n",
       "B.cell__ACTB  -0.21732116  0.18824680 -0.11174294  0.080270097 -0.06805549\n",
       "B.cell__ACTG1 -0.03822634 -0.02855239 -0.04507595  0.062922463 -0.05201821\n",
       "B.cell__ACTR2 -0.08317667  0.21396732 -0.16865762  0.004320409  0.02414656\n",
       "B.cell__ACTR3  0.07242947  0.04114458 -0.17622976 -0.093380645 -0.02262458\n",
       "B.cell__ADD3  -0.10174651  0.14137915 -0.18193237 -0.028330170 -0.03031231\n",
       "B.cell__AES    0.18231821 -0.12558725  0.11752542  0.090172626 -0.06976022\n",
       "              Factor6       Factor7      Factor8    Factor9      Factor10     ⋯\n",
       "B.cell__ACTB  -0.0421615392 -0.064257897 -0.4432788  0.007395637  0.034436032 ⋯\n",
       "B.cell__ACTG1 -0.0592270233  0.067477025 -0.3285550 -0.012521170  0.030991051 ⋯\n",
       "B.cell__ACTR2  0.0572508350  0.007391086 -0.2363539  0.001659897  0.261096179 ⋯\n",
       "B.cell__ACTR3  0.0004040518  0.008664441 -0.1160945  0.010067453  0.024094608 ⋯\n",
       "B.cell__ADD3  -0.0162211861 -0.005273766 -0.2519083 -0.013892147 -0.003902171 ⋯\n",
       "B.cell__AES    0.0765330419  0.073054507 -0.1399591 -0.022372559 -0.010901893 ⋯\n",
       "              Factor15    Factor16    Factor17    Factor18    Factor19   \n",
       "B.cell__ACTB   0.01741151  0.01803296  0.07789122  0.03405593 -0.14379902\n",
       "B.cell__ACTG1 -0.02331705  0.03813000  0.10040610  0.10898526  0.02983564\n",
       "B.cell__ACTR2  0.02130334  0.02949983 -0.06087974  0.02812659 -0.13211621\n",
       "B.cell__ACTR3 -0.08647003  0.00610388 -0.05120527  0.05020532  0.02533577\n",
       "B.cell__ADD3  -0.11196484 -0.02243120  0.06199035 -0.02736166  0.04897216\n",
       "B.cell__AES    0.12049376  0.02960805  0.04202096  0.02239864 -0.05132128\n",
       "              Factor20    type   variable_name view   gene \n",
       "B.cell__ACTB  -0.03069997 B.cell B.cell__ACTB  B.cell ACTB \n",
       "B.cell__ACTG1 -0.14744236 B.cell B.cell__ACTG1 B.cell ACTG1\n",
       "B.cell__ACTR2 -0.17683296 B.cell B.cell__ACTR2 B.cell ACTR2\n",
       "B.cell__ACTR3 -0.09245851 B.cell B.cell__ACTR3 B.cell ACTR3\n",
       "B.cell__ADD3  -0.04420387 B.cell B.cell__ADD3  B.cell ADD3 \n",
       "B.cell__AES   -0.26910231 B.cell B.cell__AES   B.cell AES  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "weights = get_weights(model, views = \"all\", factors = \"all\")\n",
    "weight_data = data.frame()\n",
    "for (i in names(weights)){\n",
    "    data = data.frame(weights[[i]])\n",
    "    data$type = i\n",
    "    weight_data = rbind(weight_data,data)\n",
    "}\n",
    "weight_data$variable_name = rownames(weight_data)\n",
    "weight_data$view <- weight_data$type \n",
    "weight_data$gene = sapply(strsplit(weight_data$variable_name, \"_\"), \"[\", 3)\n",
    "head(weight_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "11afe383-a846-4e74-9b96-af6bd0043b15",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Transform to long format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d7769bc7-2dab-43b3-ab1d-ca7943e1afc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using type, variable_name, view, gene as id variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "feature_weights_long = melt(weight_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2f2f453f-0550-47c6-9cde-12498561987d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>type</th><th scope=col>variable_name</th><th scope=col>view</th><th scope=col>gene</th><th scope=col>variable</th><th scope=col>value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.cell</td><td>B.cell__ACTB </td><td>B.cell</td><td>ACTB </td><td>Factor1</td><td>-0.21732116</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>B.cell</td><td>B.cell__ACTG1</td><td>B.cell</td><td>ACTG1</td><td>Factor1</td><td>-0.03822634</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & type & variable\\_name & view & gene & variable & value\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & B.cell & B.cell\\_\\_ACTB  & B.cell & ACTB  & Factor1 & -0.21732116\\\\\n",
       "\t2 & B.cell & B.cell\\_\\_ACTG1 & B.cell & ACTG1 & Factor1 & -0.03822634\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | type &lt;chr&gt; | variable_name &lt;chr&gt; | view &lt;chr&gt; | gene &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | B.cell | B.cell__ACTB  | B.cell | ACTB  | Factor1 | -0.21732116 |\n",
       "| 2 | B.cell | B.cell__ACTG1 | B.cell | ACTG1 | Factor1 | -0.03822634 |\n",
       "\n"
      ],
      "text/plain": [
       "  type   variable_name view   gene  variable value      \n",
       "1 B.cell B.cell__ACTB  B.cell ACTB  Factor1  -0.21732116\n",
       "2 B.cell B.cell__ACTG1 B.cell ACTG1 Factor1  -0.03822634"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_weights_long,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "051807fc-3e5b-4262-977a-70739f505aff",
   "metadata": {},
   "source": [
    "## Adjust feature names in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "50d2a2a2-7e4c-4eb6-89c4-653ded4e69b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B.cell__ACTB'</li><li>'B.cell__ACTG1'</li><li>'B.cell__ACTR2'</li><li>'B.cell__ACTR3'</li><li>'B.cell__ADD3'</li><li>'B.cell__AES'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B.cell\\_\\_ACTB'\n",
       "\\item 'B.cell\\_\\_ACTG1'\n",
       "\\item 'B.cell\\_\\_ACTR2'\n",
       "\\item 'B.cell\\_\\_ACTR3'\n",
       "\\item 'B.cell\\_\\_ADD3'\n",
       "\\item 'B.cell\\_\\_AES'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B.cell__ACTB'\n",
       "2. 'B.cell__ACTG1'\n",
       "3. 'B.cell__ACTR2'\n",
       "4. 'B.cell__ACTR3'\n",
       "5. 'B.cell__ADD3'\n",
       "6. 'B.cell__AES'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B.cell__ACTB\"  \"B.cell__ACTG1\" \"B.cell__ACTR2\" \"B.cell__ACTR3\"\n",
       "[5] \"B.cell__ADD3\"  \"B.cell__AES\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(features_names(model)[[1]] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "790decbc-1976-4fce-b491-5c80b4eca02b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_conc = model ## save original model for later\n",
    "for(i in names(features_names(model))){   \n",
    "        features_names(model)[[i]] = with(feature_weights_long, gene[match(features_names(model)[[i]], variable_name)])\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7ce46cdd-57a2-474e-8164-2d43494d9d88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ACTB'</li><li>'ACTG1'</li><li>'ACTR2'</li><li>'ACTR3'</li><li>'ADD3'</li><li>'AES'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ACTB'\n",
       "\\item 'ACTG1'\n",
       "\\item 'ACTR2'\n",
       "\\item 'ACTR3'\n",
       "\\item 'ADD3'\n",
       "\\item 'AES'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ACTB'\n",
       "2. 'ACTG1'\n",
       "3. 'ACTR2'\n",
       "4. 'ACTR3'\n",
       "5. 'ADD3'\n",
       "6. 'AES'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ACTB\"  \"ACTG1\" \"ACTR2\" \"ACTR3\" \"ADD3\"  \"AES\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(features_names(model)[[1]] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68f118e7-c714-4e92-9b4a-bb00c10008e5",
   "metadata": {},
   "source": [
    "## Create a MOFA model for the overall enrichment analysis across views"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e16a1da5-3d8c-4045-9be0-676bdad5d9d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need a model that has features of all views concatenated in a single view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "cae7d861-6003-4ec4-91e2-c1eed7a851be",
   "metadata": {},
   "outputs": [],
   "source": [
    "views <- names(features_names(model_conc))\n",
    "tmp <- sapply(views, function(view) model_conc@intercepts[[view]]$group1)\n",
    "names(tmp) <- NULL  \n",
    "model_conc@intercepts[['complete']]$group1 = unlist(tmp)\n",
    "model_conc@expectations$W[['complete']] = do.call(rbind, lapply(views, function(view) model_conc@expectations$W[[view]]))\n",
    "model_conc@features_metadata = rbind(model_conc@features_metadata, \n",
    "                                     data.frame(feature = model_conc@features_metadata$feature, view=\"complete\"))\n",
    "model_conc@dimensions$D['complete'] = sum(model_conc@dimensions$D)\n",
    "model_conc@data[['complete']]$group1 = do.call(rbind, lapply(views, function(view) model_conc@data[[view]]$group1))\n",
    "model_conc@data_options$views = c(model_conc@data_options$views , 'complete')\n",
    "model_conc@model_options$likelihoods['complete'] = 'gaussian'\n",
    "model_conc@dimensions$M = length(views) + 1\n",
    "\n",
    "model_conc@cache$variance_explained$r2_total$group1['complete'] = mean(model_conc@cache$variance_explained$r2_total$group1)\n",
    "model_conc@cache$variance_explained$r2_per_factor$group1 = cbind(model@cache$variance_explained$r2_per_factor$group1, data.frame('complete' = rowMeans(model_conc@cache$variance_explained$r2_per_factor$group1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "62640a96-f3c2-47ae-98a1-75d58c0a0285",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Trained MOFA with the following characteristics: \n",
       " Number of views: 18 \n",
       " Views names: B.cell CD14.Mono CD16.Mono CD4.CTL CD4.Naive CD4.TCM CD4.TEM CD8.Naive CD8.TEM cDC2 clinical cytokine gdT neutrophil NK proteomic Treg complete \n",
       " Number of features (per view): 686 1589 1059 542 473 661 737 397 570 280 4 66 218 981 508 490 364 9625 \n",
       " Number of groups: 1 \n",
       " Groups names: group1 \n",
       " Number of samples (per group): 128 \n",
       " Number of factors: 20 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "model_conc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b2766b1-3128-45b2-9719-6d883a46d729",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fa56b7a6-ed4d-47e8-9efa-cc660d8e25ae",
   "metadata": {},
   "source": [
    "# Prepare pathway data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1f307813-471a-4643-aac5-88f9f1a5d5fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>gene</th><th scope=col>pathway_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-1059683</td><td>JAK1 </td><td>Interleukin-6 signaling</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>STAT1</td><td>Interleukin-6 signaling</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & ID & gene & pathway\\_name\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-1059683 & JAK1  & Interleukin-6 signaling\\\\\n",
       "\t2 & R-HSA-1059683 & STAT1 & Interleukin-6 signaling\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | gene &lt;chr&gt; | pathway_name &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | R-HSA-1059683 | JAK1  | Interleukin-6 signaling |\n",
       "| 2 | R-HSA-1059683 | STAT1 | Interleukin-6 signaling |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            gene  pathway_name           \n",
       "1 R-HSA-1059683 JAK1  Interleukin-6 signaling\n",
       "2 R-HSA-1059683 STAT1 Interleukin-6 signaling"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(pathways,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80fe97f2-6c73-441a-bcb7-74ea08223baa",
   "metadata": {},
   "source": [
    "## Filter pathways out based on coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6d4a622f-bdfc-4868-a36a-3dedd09c143c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## get MOFA input genes \n",
    "mofa_genes = data.frame(gene = unique(feature_weights_long$gene), is_feature = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1b52630c-66cb-4d1f-a7d4-a323a0130f17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>is_feature</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ACTB </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ACTG1</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene & is\\_feature\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ACTB  & 1\\\\\n",
       "\t2 & ACTG1 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | is_feature &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ACTB  | 1 |\n",
       "| 2 | ACTG1 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  is_feature\n",
       "1 ACTB  1         \n",
       "2 ACTG1 1         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mofa_genes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d5a1ffd9-1884-4885-8994-1e0bc28fc8ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge features within mofa model to pathways\n",
    "feature_set = merge(pathways, mofa_genes, all.x = TRUE)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "eb6f38e6-336f-4ac1-8fae-ab70f6672ea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>is_feature</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>R-HSA-168256 </td><td>Immune System           </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A1BG</td><td>R-HSA-6798695</td><td>Neutrophil degranulation</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & gene & ID & pathway\\_name & is\\_feature\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & R-HSA-168256  & Immune System            & 1\\\\\n",
       "\t2 & A1BG & R-HSA-6798695 & Neutrophil degranulation & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 4\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | ID &lt;chr&gt; | pathway_name &lt;chr&gt; | is_feature &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | A1BG | R-HSA-168256  | Immune System            | 1 |\n",
       "| 2 | A1BG | R-HSA-6798695 | Neutrophil degranulation | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene ID            pathway_name             is_feature\n",
       "1 A1BG R-HSA-168256  Immune System            1         \n",
       "2 A1BG R-HSA-6798695 Neutrophil degranulation 1         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1b48c474-4c44-40d7-bc4a-f781e4f050e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'ID'. You can override using the `.groups` argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'ID'. You can override using the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "137"
      ],
      "text/latex": [
       "137"
      ],
      "text/markdown": [
       "137"
      ],
      "text/plain": [
       "[1] 137"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Remove pathways for which we have not a high amount of genes in our data (coverage_par)\n",
    "filter = feature_set %>% group_by(ID, pathway_name) %>% summarise(gene_amount = n(),matched_amount = sum(!is.na(is_feature)),  coverage = sum(!is.na(is_feature)) / n()) %>% filter(coverage >=  coverage_par)\n",
    "\n",
    "### Get the pathways that have been filtered out because of to low coverage\n",
    "filtered_pathways = feature_set %>% group_by(ID, pathway_name) %>% summarise(gene_amount = n(),matched_amount = sum(!is.na(is_feature)),  coverage = sum(!is.na(is_feature)) / n()) %>% filter(coverage <  coverage_par)\n",
    "nrow(filtered_pathways)  # amount of pathways that have een filtred out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "75d6214a-da9e-45c2-be09-5fb2bfcb12e1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "feature_set = merge(feature_set, filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "e8d5e187-ecc3-4f86-97f4-d86afa5888af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 15 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>gene_amount</th><th scope=col>matched_amount</th><th scope=col>coverage</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>R-HSA-1168372</td><td>Downstream signaling events of B Cell Receptor (BCR)        </td><td>101</td><td> 39</td><td>0.3861386</td></tr>\n",
       "\t<tr><td>R-HSA-1169091</td><td>Activation of NF-kappaB in B cells                          </td><td> 73</td><td> 25</td><td>0.3424658</td></tr>\n",
       "\t<tr><td>R-HSA-1169092</td><td>Activation of RAS in B cells                                </td><td>  7</td><td>  3</td><td>0.4285714</td></tr>\n",
       "\t<tr><td>R-HSA-1169408</td><td>ISG15 antiviral mechanism                                   </td><td> 72</td><td> 22</td><td>0.3055556</td></tr>\n",
       "\t<tr><td>R-HSA-1169410</td><td>Antiviral mechanism by IFN-stimulated genes                 </td><td> 80</td><td> 23</td><td>0.2875000</td></tr>\n",
       "\t<tr><td>R-HSA-1170546</td><td>Prolactin receptor signaling                                </td><td> 21</td><td>  3</td><td>0.1428571</td></tr>\n",
       "\t<tr><td>R-HSA-1222556</td><td>ROS and RNS production in phagocytes                        </td><td> 36</td><td> 16</td><td>0.4444444</td></tr>\n",
       "\t<tr><td>R-HSA-1236974</td><td>ER-Phagosome pathway                                        </td><td>100</td><td> 46</td><td>0.4600000</td></tr>\n",
       "\t<tr><td>R-HSA-1236975</td><td>Antigen processing-Cross presentation                       </td><td>164</td><td> 70</td><td>0.4268293</td></tr>\n",
       "\t<tr><td>R-HSA-1236977</td><td>Endosomal/Vacuolar pathway                                  </td><td> 11</td><td>  3</td><td>0.2727273</td></tr>\n",
       "\t<tr><td>R-HSA-1236978</td><td>Cross-presentation of soluble exogenous antigens (endosomes)</td><td> 53</td><td> 17</td><td>0.3207547</td></tr>\n",
       "\t<tr><td>R-HSA-1266695</td><td>Interleukin-7 signaling                                     </td><td> 63</td><td> 19</td><td>0.3015873</td></tr>\n",
       "\t<tr><td>R-HSA-1280215</td><td>Cytokine Signaling in Immune system                         </td><td>824</td><td>280</td><td>0.3398058</td></tr>\n",
       "\t<tr><td>R-HSA-1280218</td><td>Adaptive Immune System                                      </td><td>877</td><td>263</td><td>0.2998860</td></tr>\n",
       "\t<tr><td>R-HSA-1461957</td><td>Beta defensins                                              </td><td> 40</td><td>  2</td><td>0.0500000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 15 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " ID & pathway\\_name & gene\\_amount & matched\\_amount & coverage\\\\\n",
       " <chr> & <chr> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t R-HSA-1168372 & Downstream signaling events of B Cell Receptor (BCR)         & 101 &  39 & 0.3861386\\\\\n",
       "\t R-HSA-1169091 & Activation of NF-kappaB in B cells                           &  73 &  25 & 0.3424658\\\\\n",
       "\t R-HSA-1169092 & Activation of RAS in B cells                                 &   7 &   3 & 0.4285714\\\\\n",
       "\t R-HSA-1169408 & ISG15 antiviral mechanism                                    &  72 &  22 & 0.3055556\\\\\n",
       "\t R-HSA-1169410 & Antiviral mechanism by IFN-stimulated genes                  &  80 &  23 & 0.2875000\\\\\n",
       "\t R-HSA-1170546 & Prolactin receptor signaling                                 &  21 &   3 & 0.1428571\\\\\n",
       "\t R-HSA-1222556 & ROS and RNS production in phagocytes                         &  36 &  16 & 0.4444444\\\\\n",
       "\t R-HSA-1236974 & ER-Phagosome pathway                                         & 100 &  46 & 0.4600000\\\\\n",
       "\t R-HSA-1236975 & Antigen processing-Cross presentation                        & 164 &  70 & 0.4268293\\\\\n",
       "\t R-HSA-1236977 & Endosomal/Vacuolar pathway                                   &  11 &   3 & 0.2727273\\\\\n",
       "\t R-HSA-1236978 & Cross-presentation of soluble exogenous antigens (endosomes) &  53 &  17 & 0.3207547\\\\\n",
       "\t R-HSA-1266695 & Interleukin-7 signaling                                      &  63 &  19 & 0.3015873\\\\\n",
       "\t R-HSA-1280215 & Cytokine Signaling in Immune system                          & 824 & 280 & 0.3398058\\\\\n",
       "\t R-HSA-1280218 & Adaptive Immune System                                       & 877 & 263 & 0.2998860\\\\\n",
       "\t R-HSA-1461957 & Beta defensins                                               &  40 &   2 & 0.0500000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 15 × 5\n",
       "\n",
       "| ID &lt;chr&gt; | pathway_name &lt;chr&gt; | gene_amount &lt;int&gt; | matched_amount &lt;int&gt; | coverage &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| R-HSA-1168372 | Downstream signaling events of B Cell Receptor (BCR)         | 101 |  39 | 0.3861386 |\n",
       "| R-HSA-1169091 | Activation of NF-kappaB in B cells                           |  73 |  25 | 0.3424658 |\n",
       "| R-HSA-1169092 | Activation of RAS in B cells                                 |   7 |   3 | 0.4285714 |\n",
       "| R-HSA-1169408 | ISG15 antiviral mechanism                                    |  72 |  22 | 0.3055556 |\n",
       "| R-HSA-1169410 | Antiviral mechanism by IFN-stimulated genes                  |  80 |  23 | 0.2875000 |\n",
       "| R-HSA-1170546 | Prolactin receptor signaling                                 |  21 |   3 | 0.1428571 |\n",
       "| R-HSA-1222556 | ROS and RNS production in phagocytes                         |  36 |  16 | 0.4444444 |\n",
       "| R-HSA-1236974 | ER-Phagosome pathway                                         | 100 |  46 | 0.4600000 |\n",
       "| R-HSA-1236975 | Antigen processing-Cross presentation                        | 164 |  70 | 0.4268293 |\n",
       "| R-HSA-1236977 | Endosomal/Vacuolar pathway                                   |  11 |   3 | 0.2727273 |\n",
       "| R-HSA-1236978 | Cross-presentation of soluble exogenous antigens (endosomes) |  53 |  17 | 0.3207547 |\n",
       "| R-HSA-1266695 | Interleukin-7 signaling                                      |  63 |  19 | 0.3015873 |\n",
       "| R-HSA-1280215 | Cytokine Signaling in Immune system                          | 824 | 280 | 0.3398058 |\n",
       "| R-HSA-1280218 | Adaptive Immune System                                       | 877 | 263 | 0.2998860 |\n",
       "| R-HSA-1461957 | Beta defensins                                               |  40 |   2 | 0.0500000 |\n",
       "\n"
      ],
      "text/plain": [
       "   ID            pathway_name                                                \n",
       "1  R-HSA-1168372 Downstream signaling events of B Cell Receptor (BCR)        \n",
       "2  R-HSA-1169091 Activation of NF-kappaB in B cells                          \n",
       "3  R-HSA-1169092 Activation of RAS in B cells                                \n",
       "4  R-HSA-1169408 ISG15 antiviral mechanism                                   \n",
       "5  R-HSA-1169410 Antiviral mechanism by IFN-stimulated genes                 \n",
       "6  R-HSA-1170546 Prolactin receptor signaling                                \n",
       "7  R-HSA-1222556 ROS and RNS production in phagocytes                        \n",
       "8  R-HSA-1236974 ER-Phagosome pathway                                        \n",
       "9  R-HSA-1236975 Antigen processing-Cross presentation                       \n",
       "10 R-HSA-1236977 Endosomal/Vacuolar pathway                                  \n",
       "11 R-HSA-1236978 Cross-presentation of soluble exogenous antigens (endosomes)\n",
       "12 R-HSA-1266695 Interleukin-7 signaling                                     \n",
       "13 R-HSA-1280215 Cytokine Signaling in Immune system                         \n",
       "14 R-HSA-1280218 Adaptive Immune System                                      \n",
       "15 R-HSA-1461957 Beta defensins                                              \n",
       "   gene_amount matched_amount coverage \n",
       "1  101          39            0.3861386\n",
       "2   73          25            0.3424658\n",
       "3    7           3            0.4285714\n",
       "4   72          22            0.3055556\n",
       "5   80          23            0.2875000\n",
       "6   21           3            0.1428571\n",
       "7   36          16            0.4444444\n",
       "8  100          46            0.4600000\n",
       "9  164          70            0.4268293\n",
       "10  11           3            0.2727273\n",
       "11  53          17            0.3207547\n",
       "12  63          19            0.3015873\n",
       "13 824         280            0.3398058\n",
       "14 877         263            0.2998860\n",
       "15  40           2            0.0500000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Overview pathways that have been excluded from testing due to low amount of matching genes\n",
    "head(filtered_pathways,15) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c3014e7c-342a-451e-94de-853f50acd626",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove NA entries for not mapped genes in feature set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "146410de-7428-47e2-bd55-49f0b98ff3bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>gene</th><th scope=col>is_feature</th><th scope=col>gene_amount</th><th scope=col>matched_amount</th><th scope=col>coverage</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>CBL  </td><td>NA</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>STAT1</td><td> 1</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & ID & pathway\\_name & gene & is\\_feature & gene\\_amount & matched\\_amount & coverage\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-1059683 & Interleukin-6 signaling & CBL   & NA & 15 & 11 & 0.7333333\\\\\n",
       "\t2 & R-HSA-1059683 & Interleukin-6 signaling & STAT1 &  1 & 15 & 11 & 0.7333333\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | pathway_name &lt;chr&gt; | gene &lt;chr&gt; | is_feature &lt;dbl&gt; | gene_amount &lt;int&gt; | matched_amount &lt;int&gt; | coverage &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | R-HSA-1059683 | Interleukin-6 signaling | CBL   | NA | 15 | 11 | 0.7333333 |\n",
       "| 2 | R-HSA-1059683 | Interleukin-6 signaling | STAT1 |  1 | 15 | 11 | 0.7333333 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            pathway_name            gene  is_feature gene_amount\n",
       "1 R-HSA-1059683 Interleukin-6 signaling CBL   NA         15         \n",
       "2 R-HSA-1059683 Interleukin-6 signaling STAT1  1         15         \n",
       "  matched_amount coverage \n",
       "1 11             0.7333333\n",
       "2 11             0.7333333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b2adb74e-debb-4cb3-850e-c68f395b1660",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set = feature_set[!is.na(feature_set$is_feature),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1df77c8e-4ca0-417b-82ed-596f4da1f27b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1100"
      ],
      "text/latex": [
       "1100"
      ],
      "text/markdown": [
       "1100"
      ],
      "text/plain": [
       "[1] 1100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(feature_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "9ae60ad7-c0d2-4d23-92bb-2728d0aea93d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>gene</th><th scope=col>is_feature</th><th scope=col>gene_amount</th><th scope=col>matched_amount</th><th scope=col>coverage</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>STAT1</td><td>1</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>SOCS3</td><td>1</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & ID & pathway\\_name & gene & is\\_feature & gene\\_amount & matched\\_amount & coverage\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t2 & R-HSA-1059683 & Interleukin-6 signaling & STAT1 & 1 & 15 & 11 & 0.7333333\\\\\n",
       "\t3 & R-HSA-1059683 & Interleukin-6 signaling & SOCS3 & 1 & 15 & 11 & 0.7333333\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | pathway_name &lt;chr&gt; | gene &lt;chr&gt; | is_feature &lt;dbl&gt; | gene_amount &lt;int&gt; | matched_amount &lt;int&gt; | coverage &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 2 | R-HSA-1059683 | Interleukin-6 signaling | STAT1 | 1 | 15 | 11 | 0.7333333 |\n",
       "| 3 | R-HSA-1059683 | Interleukin-6 signaling | SOCS3 | 1 | 15 | 11 | 0.7333333 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            pathway_name            gene  is_feature gene_amount\n",
       "2 R-HSA-1059683 Interleukin-6 signaling STAT1 1          15         \n",
       "3 R-HSA-1059683 Interleukin-6 signaling SOCS3 1          15         \n",
       "  matched_amount coverage \n",
       "2 11             0.7333333\n",
       "3 11             0.7333333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2c07ab49-dfd9-4f6f-861a-cbfc8364fcd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save coverage for later usage\n",
    "coverage_info = unique(feature_set[,c('ID', 'pathway_name', 'coverage')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98400d4f-3ec8-40df-8c4d-dc74bda48e73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b0b1ae77-74ea-4c54-ae66-50743120a01e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get genes that are not mapped / and add a not mapped pathway to use them in the background later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "935abc0f-9ee2-4d15-a711-7267c98e3847",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_pathway_genes = unique(mofa_genes$gene[!mofa_genes$gene %in% unique(feature_set$gene)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "db769bb8-f36e-4b66-ada1-823c6943c6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_pathway = data.frame(ID = 'Background', pathway_name = 'Background', gene = non_pathway_genes, \n",
    "                         is_feature = 1, gene_amount = length(non_pathway_genes), matched_amount =  length(non_pathway_genes),\n",
    "                         coverage = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "3dcaf535-ee54-4a3c-b9fa-91e6639e6ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set = rbind(feature_set, add_pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c29eae3-ba43-4960-82ae-3ce710d83414",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1e7e623b-f046-47c8-ba44-76e2503c7abc",
   "metadata": {},
   "source": [
    "## Transform to binary matrix format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1ac2c0be-f5ae-40d7-a5c0-8471f76921aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>gene</th><th scope=col>is_feature</th><th scope=col>gene_amount</th><th scope=col>matched_amount</th><th scope=col>coverage</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>STAT1</td><td>1</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>SOCS3</td><td>1</td><td>15</td><td>11</td><td>0.7333333</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & ID & pathway\\_name & gene & is\\_feature & gene\\_amount & matched\\_amount & coverage\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t2 & R-HSA-1059683 & Interleukin-6 signaling & STAT1 & 1 & 15 & 11 & 0.7333333\\\\\n",
       "\t3 & R-HSA-1059683 & Interleukin-6 signaling & SOCS3 & 1 & 15 & 11 & 0.7333333\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | pathway_name &lt;chr&gt; | gene &lt;chr&gt; | is_feature &lt;dbl&gt; | gene_amount &lt;int&gt; | matched_amount &lt;int&gt; | coverage &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 2 | R-HSA-1059683 | Interleukin-6 signaling | STAT1 | 1 | 15 | 11 | 0.7333333 |\n",
       "| 3 | R-HSA-1059683 | Interleukin-6 signaling | SOCS3 | 1 | 15 | 11 | 0.7333333 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            pathway_name            gene  is_feature gene_amount\n",
       "2 R-HSA-1059683 Interleukin-6 signaling STAT1 1          15         \n",
       "3 R-HSA-1059683 Interleukin-6 signaling SOCS3 1          15         \n",
       "  matched_amount coverage \n",
       "2 11             0.7333333\n",
       "3 11             0.7333333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d7b37e6d-f59f-42fe-b287-2c7760bb147b",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set$pathway_id = paste0(feature_set$ID, '_', feature_set$pathway_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3a65c879-1c8b-4483-bf6f-867d6c41529f",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set$value = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "422b00a6-56c2-4702-8785-b585a9bc2cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Adjust names for overall approach (concatenate gene names with cell-types and match to pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "93eb6290-2d0a-49fe-b1f5-fc14e79b5219",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_weights_long_mapped_join = unique(feature_weights_long[,c('gene', 'variable_name')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "db7b5b36-c929-4cfc-9c07-2cb89b2c6fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>variable_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ACTB </td><td>B.cell__ACTB </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ACTG1</td><td>B.cell__ACTG1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene & variable\\_name\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ACTB  & B.cell\\_\\_ACTB \\\\\n",
       "\t2 & ACTG1 & B.cell\\_\\_ACTG1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | variable_name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ACTB  | B.cell__ACTB  |\n",
       "| 2 | ACTG1 | B.cell__ACTG1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  variable_name\n",
       "1 ACTB  B.cell__ACTB \n",
       "2 ACTG1 B.cell__ACTG1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_weights_long_mapped_join,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c3341be4-994f-4090-8b67-4a8f6623dfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set_all = merge(feature_set, feature_weights_long_mapped_join, by.x = 'gene', by.y = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "115be6a6-de7c-4f86-8d37-67399d5038e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "12263"
      ],
      "text/latex": [
       "12263"
      ],
      "text/markdown": [
       "12263"
      ],
      "text/plain": [
       "[1] 12263"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(feature_set_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d2685fc3-48d5-48bd-8d44-2748496d3750",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Binarize feature set to use in type-specific enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b0dd4fff-470c-498d-95dd-c71acccb7ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Based on gene-name for overall enrichment\n",
    "\n",
    "feature_set = unique(feature_set[,c('pathway_id', 'gene', 'value')]) %>% dcast(pathway_id ~ gene, value.var = 'value')\n",
    "feature_set[is.na(feature_set)]= 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "44e7a0cc-99c6-4f31-be40-41854ea44927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2677</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway_id</th><th scope=col>A1BG</th><th scope=col>A2M</th><th scope=col>AAK1</th><th scope=col>AATK</th><th scope=col>ABCA7</th><th scope=col>ABHD17A</th><th scope=col>ABHD2</th><th scope=col>ABHD5</th><th scope=col>ABI1</th><th scope=col>⋯</th><th scope=col>ZNF331</th><th scope=col>ZNF385A</th><th scope=col>ZNF467</th><th scope=col>ZNF511.PRAP1</th><th scope=col>ZNF706</th><th scope=col>ZNFX1</th><th scope=col>ZNHIT1</th><th scope=col>ZRANB2</th><th scope=col>ZSWIM6</th><th scope=col>ZYX</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Background_Background                </td><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>0</td><td>⋯</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683_Interleukin-6 signaling</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2677\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & pathway\\_id & A1BG & A2M & AAK1 & AATK & ABCA7 & ABHD17A & ABHD2 & ABHD5 & ABI1 & ⋯ & ZNF331 & ZNF385A & ZNF467 & ZNF511.PRAP1 & ZNF706 & ZNFX1 & ZNHIT1 & ZRANB2 & ZSWIM6 & ZYX\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & Background\\_Background                 & 0 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 0 & ⋯ & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1 & 1\\\\\n",
       "\t2 & R-HSA-1059683\\_Interleukin-6 signaling & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2677\n",
       "\n",
       "| <!--/--> | pathway_id &lt;chr&gt; | A1BG &lt;dbl&gt; | A2M &lt;dbl&gt; | AAK1 &lt;dbl&gt; | AATK &lt;dbl&gt; | ABCA7 &lt;dbl&gt; | ABHD17A &lt;dbl&gt; | ABHD2 &lt;dbl&gt; | ABHD5 &lt;dbl&gt; | ABI1 &lt;dbl&gt; | ⋯ ⋯ | ZNF331 &lt;dbl&gt; | ZNF385A &lt;dbl&gt; | ZNF467 &lt;dbl&gt; | ZNF511.PRAP1 &lt;dbl&gt; | ZNF706 &lt;dbl&gt; | ZNFX1 &lt;dbl&gt; | ZNHIT1 &lt;dbl&gt; | ZRANB2 &lt;dbl&gt; | ZSWIM6 &lt;dbl&gt; | ZYX &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Background_Background                 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0 | ⋯ | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 |\n",
       "| 2 | R-HSA-1059683_Interleukin-6 signaling | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway_id                            A1BG A2M AAK1 AATK ABCA7 ABHD17A ABHD2\n",
       "1 Background_Background                 0    1   1    1    1     1       1    \n",
       "2 R-HSA-1059683_Interleukin-6 signaling 0    0   0    0    0     0       0    \n",
       "  ABHD5 ABI1 ⋯ ZNF331 ZNF385A ZNF467 ZNF511.PRAP1 ZNF706 ZNFX1 ZNHIT1 ZRANB2\n",
       "1 1     0    ⋯ 1      1       1      1            1      1     1      1     \n",
       "2 0     0    ⋯ 0      0       0      0            0      0     0      0     \n",
       "  ZSWIM6 ZYX\n",
       "1 1      1  \n",
       "2 0      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "c3acbc39-84e9-4970-bb45-65f954bf6c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Binarize feature set to use in overall enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0e9efb5b-139e-447a-b68a-16c6a227e6ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set_all = unique(feature_set_all[,c('pathway_id', 'variable_name', 'value')]) %>% dcast(pathway_id ~ variable_name, value.var = 'value')\n",
    "feature_set_all[is.na(feature_set_all)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d0cb106a-277c-406b-a503-083733148b77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9626</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway_id</th><th scope=col>B.cell__ACTB</th><th scope=col>B.cell__ACTG1</th><th scope=col>B.cell__ACTR2</th><th scope=col>B.cell__ACTR3</th><th scope=col>B.cell__ADD3</th><th scope=col>B.cell__AES</th><th scope=col>B.cell__AFF3</th><th scope=col>B.cell__AHNAK</th><th scope=col>B.cell__AKAP13</th><th scope=col>⋯</th><th scope=col>Treg__VIM</th><th scope=col>Treg__YBX1</th><th scope=col>Treg__YPEL5</th><th scope=col>Treg__YWHAB</th><th scope=col>Treg__YWHAZ</th><th scope=col>Treg__ZC3HAV1</th><th scope=col>Treg__ZFAS1</th><th scope=col>Treg__ZFP36</th><th scope=col>Treg__ZFP36L2</th><th scope=col>Treg__ZNF331</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Background_Background                </td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>⋯</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683_Interleukin-6 signaling</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9626\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & pathway\\_id & B.cell\\_\\_ACTB & B.cell\\_\\_ACTG1 & B.cell\\_\\_ACTR2 & B.cell\\_\\_ACTR3 & B.cell\\_\\_ADD3 & B.cell\\_\\_AES & B.cell\\_\\_AFF3 & B.cell\\_\\_AHNAK & B.cell\\_\\_AKAP13 & ⋯ & Treg\\_\\_VIM & Treg\\_\\_YBX1 & Treg\\_\\_YPEL5 & Treg\\_\\_YWHAB & Treg\\_\\_YWHAZ & Treg\\_\\_ZC3HAV1 & Treg\\_\\_ZFAS1 & Treg\\_\\_ZFP36 & Treg\\_\\_ZFP36L2 & Treg\\_\\_ZNF331\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & Background\\_Background                 & 0 & 0 & 0 & 0 & 1 & 1 & 1 & 1 & 1 & ⋯ & 0 & 1 & 0 & 0 & 0 & 1 & 1 & 1 & 1 & 1\\\\\n",
       "\t2 & R-HSA-1059683\\_Interleukin-6 signaling & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9626\n",
       "\n",
       "| <!--/--> | pathway_id &lt;chr&gt; | B.cell__ACTB &lt;dbl&gt; | B.cell__ACTG1 &lt;dbl&gt; | B.cell__ACTR2 &lt;dbl&gt; | B.cell__ACTR3 &lt;dbl&gt; | B.cell__ADD3 &lt;dbl&gt; | B.cell__AES &lt;dbl&gt; | B.cell__AFF3 &lt;dbl&gt; | B.cell__AHNAK &lt;dbl&gt; | B.cell__AKAP13 &lt;dbl&gt; | ⋯ ⋯ | Treg__VIM &lt;dbl&gt; | Treg__YBX1 &lt;dbl&gt; | Treg__YPEL5 &lt;dbl&gt; | Treg__YWHAB &lt;dbl&gt; | Treg__YWHAZ &lt;dbl&gt; | Treg__ZC3HAV1 &lt;dbl&gt; | Treg__ZFAS1 &lt;dbl&gt; | Treg__ZFP36 &lt;dbl&gt; | Treg__ZFP36L2 &lt;dbl&gt; | Treg__ZNF331 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Background_Background                 | 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | ⋯ | 0 | 1 | 0 | 0 | 0 | 1 | 1 | 1 | 1 | 1 |\n",
       "| 2 | R-HSA-1059683_Interleukin-6 signaling | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway_id                            B.cell__ACTB B.cell__ACTG1\n",
       "1 Background_Background                 0            0            \n",
       "2 R-HSA-1059683_Interleukin-6 signaling 0            0            \n",
       "  B.cell__ACTR2 B.cell__ACTR3 B.cell__ADD3 B.cell__AES B.cell__AFF3\n",
       "1 0             0             1            1           1           \n",
       "2 0             0             0            0           0           \n",
       "  B.cell__AHNAK B.cell__AKAP13 ⋯ Treg__VIM Treg__YBX1 Treg__YPEL5 Treg__YWHAB\n",
       "1 1             1              ⋯ 0         1          0           0          \n",
       "2 0             0              ⋯ 0         0          0           0          \n",
       "  Treg__YWHAZ Treg__ZC3HAV1 Treg__ZFAS1 Treg__ZFP36 Treg__ZFP36L2 Treg__ZNF331\n",
       "1 0           1             1           1           1             1           \n",
       "2 0           0             0           0           0             0           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(feature_set_all,2) # feature set containing only genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8c8d9a42-ccc8-4808-b69a-e48aa709dc96",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Adjust rownames and convert to matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "acc820d1-c4aa-4bd3-ae85-f9cbad2b661c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(feature_set) = feature_set$pathway_id\n",
    "rownames(feature_set_all) = feature_set_all$pathway_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d838e256-c0ab-4b74-b1aa-59eb6a73b406",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set$pathway_id = NULL\n",
    "feature_set_all$pathway_id = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "14fc0d5b-efb1-4c51-badc-1dc2325cd883",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_set = as.matrix(feature_set)\n",
    "feature_set_all = as.matrix(feature_set_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb6a3235-87fc-4125-a940-0b24214ccef6",
   "metadata": {},
   "source": [
    "# Run pathway enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f98ff41-422f-4907-9340-f43b3d14b431",
   "metadata": {},
   "source": [
    "## Per View"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "94629b40-d098-4e87-bd71-0867486d792a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Intersecting features names in the model and the gene set annotation results in a total of 686 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: B.cell \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 686 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: B.cell \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 686 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: B.cell \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 1058 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD16.Mono \n",
      "Number of feature sets: 23 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 1058 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD16.Mono \n",
      "Number of feature sets: 23 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 1058 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD16.Mono \n",
      "Number of feature sets: 23 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 542 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.CTL \n",
      "Number of feature sets: 11 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 542 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.CTL \n",
      "Number of feature sets: 11 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 542 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.CTL \n",
      "Number of feature sets: 11 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 472 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.Naive \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 472 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.Naive \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 472 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.Naive \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 660 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TCM \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 660 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TCM \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 660 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TCM \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 736 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TEM \n",
      "Number of feature sets: 14 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 736 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TEM \n",
      "Number of feature sets: 14 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“1 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 736 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: CD4.TEM \n",
      "Number of feature sets: 14 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 280 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: cDC2 \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 280 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: cDC2 \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 280 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: cDC2 \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 218 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: gdT \n",
      "Number of feature sets: 8 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 218 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: gdT \n",
      "Number of feature sets: 8 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 218 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: gdT \n",
      "Number of feature sets: 8 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 980 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: neutrophil \n",
      "Number of feature sets: 19 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 980 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: neutrophil \n",
      "Number of feature sets: 19 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 980 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: neutrophil \n",
      "Number of feature sets: 19 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“2 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 506 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: NK \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“2 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 506 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: NK \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“2 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 506 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: NK \n",
      "Number of feature sets: 13 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 364 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: Treg \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 364 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: Treg \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 364 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: Treg \n",
      "Number of feature sets: 10 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if(max(views_set) != ''){\n",
    "    enrichment_result_types = run_enrichment_pathway(\n",
    "        model = model, # MOFA Model\n",
    "        factor_set = factor_set, # list of factors for which to run the enrichment\n",
    "        views = views_set,\n",
    "        use_statistic = use_statistic, # which statistic to use\n",
    "        feature_set = feature_set, # Pathway Feature Set mapping; here use without type names\n",
    "        min_size = min_size, # Min size of genes within a pathway\n",
    "        use_test = use_test, # test used for calculating p-value\n",
    "        p_val_cutoff = p_val_cutoff, # p-value cutoff used\n",
    "        enrichment_result_p_val = data.frame())# dataset for saving result \n",
    "    } else{enrichment_result_types = ''}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ea4e8238-d4d2-4f08-aa3f-0bdd6d74d98c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#enrichment_result_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "0aa02f16-11fe-4b78-95cb-6518f71e2b4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway</th><th scope=col>variable</th><th scope=col>p</th><th scope=col>padj</th><th scope=col>view</th><th scope=col>enrichment</th><th scope=col>global_FDR</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-6798695_Neutrophil degranulation      </td><td>Factor3</td><td>4.815022e-06</td><td>5.537275e-05</td><td>CD16.Mono</td><td>negative</td><td>0.01040045</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Background_Background                       </td><td>Factor3</td><td>2.369775e-04</td><td>2.725241e-03</td><td>CD16.Mono</td><td>positive</td><td>0.18221496</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>R-HSA-6798695_Neutrophil degranulation      </td><td>Factor3</td><td>6.553155e-04</td><td>4.259551e-03</td><td>NK       </td><td>negative</td><td>0.18221496</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>R-HSA-447115_Interleukin-12 family signaling</td><td>Factor1</td><td>1.252144e-03</td><td>5.008575e-03</td><td>gdT      </td><td>all     </td><td>0.27046306</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>R-HSA-9020591_Interleukin-12 signaling      </td><td>Factor1</td><td>1.252144e-03</td><td>5.008575e-03</td><td>gdT      </td><td>all     </td><td>0.27046306</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & pathway & variable & p & padj & view & enrichment & global\\_FDR\\\\\n",
       "  & <chr> & <fct> & <dbl> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-6798695\\_Neutrophil degranulation       & Factor3 & 4.815022e-06 & 5.537275e-05 & CD16.Mono & negative & 0.01040045\\\\\n",
       "\t2 & Background\\_Background                        & Factor3 & 2.369775e-04 & 2.725241e-03 & CD16.Mono & positive & 0.18221496\\\\\n",
       "\t3 & R-HSA-6798695\\_Neutrophil degranulation       & Factor3 & 6.553155e-04 & 4.259551e-03 & NK        & negative & 0.18221496\\\\\n",
       "\t4 & R-HSA-447115\\_Interleukin-12 family signaling & Factor1 & 1.252144e-03 & 5.008575e-03 & gdT       & all      & 0.27046306\\\\\n",
       "\t5 & R-HSA-9020591\\_Interleukin-12 signaling       & Factor1 & 1.252144e-03 & 5.008575e-03 & gdT       & all      & 0.27046306\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 7\n",
       "\n",
       "| <!--/--> | pathway &lt;chr&gt; | variable &lt;fct&gt; | p &lt;dbl&gt; | padj &lt;dbl&gt; | view &lt;chr&gt; | enrichment &lt;chr&gt; | global_FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | R-HSA-6798695_Neutrophil degranulation       | Factor3 | 4.815022e-06 | 5.537275e-05 | CD16.Mono | negative | 0.01040045 |\n",
       "| 2 | Background_Background                        | Factor3 | 2.369775e-04 | 2.725241e-03 | CD16.Mono | positive | 0.18221496 |\n",
       "| 3 | R-HSA-6798695_Neutrophil degranulation       | Factor3 | 6.553155e-04 | 4.259551e-03 | NK        | negative | 0.18221496 |\n",
       "| 4 | R-HSA-447115_Interleukin-12 family signaling | Factor1 | 1.252144e-03 | 5.008575e-03 | gdT       | all      | 0.27046306 |\n",
       "| 5 | R-HSA-9020591_Interleukin-12 signaling       | Factor1 | 1.252144e-03 | 5.008575e-03 | gdT       | all      | 0.27046306 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway                                      variable p           \n",
       "1 R-HSA-6798695_Neutrophil degranulation       Factor3  4.815022e-06\n",
       "2 Background_Background                        Factor3  2.369775e-04\n",
       "3 R-HSA-6798695_Neutrophil degranulation       Factor3  6.553155e-04\n",
       "4 R-HSA-447115_Interleukin-12 family signaling Factor1  1.252144e-03\n",
       "5 R-HSA-9020591_Interleukin-12 signaling       Factor1  1.252144e-03\n",
       "  padj         view      enrichment global_FDR\n",
       "1 5.537275e-05 CD16.Mono negative   0.01040045\n",
       "2 2.725241e-03 CD16.Mono positive   0.18221496\n",
       "3 4.259551e-03 NK        negative   0.18221496\n",
       "4 5.008575e-03 gdT       all        0.27046306\n",
       "5 5.008575e-03 gdT       all        0.27046306"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if(is.data.frame(enrichment_result_types)){\n",
    "    head(enrichment_result_types  %>% arrange(padj) ,5)}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4abdba54-37fd-4b8a-9ec9-9d1545e7a98d",
   "metadata": {},
   "source": [
    "## Across Views"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "aa5ba392-04df-41d6-bbbb-5d611ff0a7a2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“7 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 9618 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: complete \n",
      "Number of feature sets: 52 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“7 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 9618 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: complete \n",
      "Number of feature sets: 52 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with negative sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Warning message in run_enrichment(model, view = i, factors = factor_set, set.statistic = c(use_statistic), :\n",
      "“7 features were removed because they had no variance in the data.\n",
      "”\n",
      "Intersecting features names in the model and the gene set annotation results in a total of 9618 features.\n",
      "\n",
      "\n",
      "Running feature set Enrichment Analysis with the following options...\n",
      "View: complete \n",
      "Number of feature sets: 52 \n",
      "Set statistic: rank.sum \n",
      "Statistical test: parametric \n",
      "\n",
      "\n",
      "Subsetting weights with positive sign\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Using pathway as id variables\n",
      "\n",
      "Using pathway as id variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "enrichment_result_all = run_enrichment_pathway(\n",
    "    model = model_conc, # MOFA Model\n",
    "    views = 'complete', # dimensions for which to run\n",
    "    factor_set = factor_set, # list of factors for which to run the enrichment\n",
    "    use_statistic = use_statistic, # which statistic to use\n",
    "    feature_set = feature_set_all, # Pathway Feature Set mapping; use the concatenated one with cell-types\n",
    "    min_size = min_size, # Min size of genes within a pathway\n",
    "    use_test = use_test, # test used for calculating p-value\n",
    "    p_val_cutoff = p_val_cutoff, # p-value cutoff used\n",
    "    enrichment_result_p_val = data.frame()) # dataset for saving results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "cc94b03a-82c3-415d-9a0c-a1c452438226",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway</th><th scope=col>variable</th><th scope=col>p</th><th scope=col>padj</th><th scope=col>view</th><th scope=col>enrichment</th><th scope=col>global_FDR</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-5686938_Regulation of TLR by endogenous ligand                           </td><td>Factor4</td><td>1.508184e-10</td><td>7.842555e-09</td><td>complete</td><td>negative</td><td>1.176383e-07</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1236973_Cross-presentation of particulate exogenous antigens (phagosomes)</td><td>Factor1</td><td>7.151990e-10</td><td>3.719035e-08</td><td>complete</td><td>positive</td><td>2.365605e-07</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>R-HSA-2029482_Regulation of actin dynamics for phagocytic cup formation        </td><td>Factor3</td><td>9.098481e-10</td><td>4.731210e-08</td><td>complete</td><td>negative</td><td>2.365605e-07</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>R-HSA-450341_Activation of the AP-1 family of transcription factors            </td><td>Factor3</td><td>3.005445e-09</td><td>7.814157e-08</td><td>complete</td><td>negative</td><td>5.860618e-07</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>R-HSA-2029480_Fcgamma receptor (FCGR) dependent phagocytosis                   </td><td>Factor3</td><td>2.228949e-08</td><td>3.863511e-07</td><td>complete</td><td>negative</td><td>3.477160e-06</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & pathway & variable & p & padj & view & enrichment & global\\_FDR\\\\\n",
       "  & <chr> & <fct> & <dbl> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-5686938\\_Regulation of TLR by endogenous ligand                            & Factor4 & 1.508184e-10 & 7.842555e-09 & complete & negative & 1.176383e-07\\\\\n",
       "\t2 & R-HSA-1236973\\_Cross-presentation of particulate exogenous antigens (phagosomes) & Factor1 & 7.151990e-10 & 3.719035e-08 & complete & positive & 2.365605e-07\\\\\n",
       "\t3 & R-HSA-2029482\\_Regulation of actin dynamics for phagocytic cup formation         & Factor3 & 9.098481e-10 & 4.731210e-08 & complete & negative & 2.365605e-07\\\\\n",
       "\t4 & R-HSA-450341\\_Activation of the AP-1 family of transcription factors             & Factor3 & 3.005445e-09 & 7.814157e-08 & complete & negative & 5.860618e-07\\\\\n",
       "\t5 & R-HSA-2029480\\_Fcgamma receptor (FCGR) dependent phagocytosis                    & Factor3 & 2.228949e-08 & 3.863511e-07 & complete & negative & 3.477160e-06\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 7\n",
       "\n",
       "| <!--/--> | pathway &lt;chr&gt; | variable &lt;fct&gt; | p &lt;dbl&gt; | padj &lt;dbl&gt; | view &lt;chr&gt; | enrichment &lt;chr&gt; | global_FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | R-HSA-5686938_Regulation of TLR by endogenous ligand                            | Factor4 | 1.508184e-10 | 7.842555e-09 | complete | negative | 1.176383e-07 |\n",
       "| 2 | R-HSA-1236973_Cross-presentation of particulate exogenous antigens (phagosomes) | Factor1 | 7.151990e-10 | 3.719035e-08 | complete | positive | 2.365605e-07 |\n",
       "| 3 | R-HSA-2029482_Regulation of actin dynamics for phagocytic cup formation         | Factor3 | 9.098481e-10 | 4.731210e-08 | complete | negative | 2.365605e-07 |\n",
       "| 4 | R-HSA-450341_Activation of the AP-1 family of transcription factors             | Factor3 | 3.005445e-09 | 7.814157e-08 | complete | negative | 5.860618e-07 |\n",
       "| 5 | R-HSA-2029480_Fcgamma receptor (FCGR) dependent phagocytosis                    | Factor3 | 2.228949e-08 | 3.863511e-07 | complete | negative | 3.477160e-06 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway                                                                        \n",
       "1 R-HSA-5686938_Regulation of TLR by endogenous ligand                           \n",
       "2 R-HSA-1236973_Cross-presentation of particulate exogenous antigens (phagosomes)\n",
       "3 R-HSA-2029482_Regulation of actin dynamics for phagocytic cup formation        \n",
       "4 R-HSA-450341_Activation of the AP-1 family of transcription factors            \n",
       "5 R-HSA-2029480_Fcgamma receptor (FCGR) dependent phagocytosis                   \n",
       "  variable p            padj         view     enrichment global_FDR  \n",
       "1 Factor4  1.508184e-10 7.842555e-09 complete negative   1.176383e-07\n",
       "2 Factor1  7.151990e-10 3.719035e-08 complete positive   2.365605e-07\n",
       "3 Factor3  9.098481e-10 4.731210e-08 complete negative   2.365605e-07\n",
       "4 Factor3  3.005445e-09 7.814157e-08 complete negative   5.860618e-07\n",
       "5 Factor3  2.228949e-08 3.863511e-07 complete negative   3.477160e-06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(enrichment_result_all %>% arrange(padj) ,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "834d67fe-fa08-4b35-975c-a786cc1ca9b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine both versions and save the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "28d752bc-a098-49c0-9805-27c347c8b3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(is.data.frame(enrichment_result_types)){\n",
    "    enrichment_result = rbind(enrichment_result_all, enrichment_result_types  )\n",
    "    } else{enrichment_result = enrichment_result_all}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "28861808-af36-4127-9145-45998c0245f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_result$ID = str_replace(enrichment_result$pathway, '_.*', '')\n",
    "enrichment_result$pathway_name = str_replace(enrichment_result$pathway, '.*_', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "5a04e82f-7b76-4c02-a645-8f09bfbae1e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_result$pathway = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "d71641e7-aee8-42ef-a2f3-5d6ebcb6f950",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_result = merge(enrichment_result, coverage_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "2cfc86d2-c727-4938-bea1-c060a1fe2a0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(enrichment_result, paste0(result_path, '/06_results/06_Pathway_Enrichment_' ,mofa_name, '.csv'), row.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18989ad9-8561-4a4f-925a-f240af0bf135",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Plot interesting pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "db021067-197b-4722-95d1-d843eec949dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Either manual pathway selection or by statistical criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b5368308-dbd7-4ae4-a3b1-c820f4aa80d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "&lt;NA&gt;"
      ],
      "text/latex": [
       "<NA>"
      ],
      "text/markdown": [
       "&lt;NA&gt;"
      ],
      "text/plain": [
       "[1] NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pathway_selection_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d4453f1c-0d3a-4144-9ea5-63d5ee2caa05",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(pathway_selection_var == '' | is.na(pathway_selection_var) ){\n",
    "    pathways_selection = enrichment_result[\n",
    "        (enrichment_result$coverage > coverage_par) & \n",
    "        (enrichment_result$enrichment == select_enrichment) &\n",
    "        (enrichment_result$global_FDR < p_value_cutoff_plot),] %>%  arrange(padj) %>% group_by(variable) %>% top_n(max_pathways, -padj)\n",
    "    } else{ pathways_selection = enrichment_result[enrichment_result$pathway_name %in% unlist(str_split(pathway_selection_var, ',')),]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48932983-7fd4-4855-8fc1-aea6fe01cb94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3218637c-9b9b-403e-ba6a-f6a76badc560",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "12b73159-f831-4cda-ad59-e99e8e07e436",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pathways_selection = unique(pathways_selection$ID)   # add factor!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "28b52d53-61d6-4b84-9857-eea067471aff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>pathway_name</th><th scope=col>variable</th><th scope=col>p</th><th scope=col>padj</th><th scope=col>view</th><th scope=col>enrichment</th><th scope=col>global_FDR</th><th scope=col>coverage</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>Factor1</td><td>1.0000000</td><td>1.0000000</td><td>complete</td><td>positive</td><td>1</td><td>0.7333333</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>R-HSA-1059683</td><td>Interleukin-6 signaling</td><td>Factor3</td><td>0.6165494</td><td>0.7286493</td><td>complete</td><td>all     </td><td>1</td><td>0.7333333</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & ID & pathway\\_name & variable & p & padj & view & enrichment & global\\_FDR & coverage\\\\\n",
       "  & <chr> & <chr> & <fct> & <dbl> & <dbl> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & R-HSA-1059683 & Interleukin-6 signaling & Factor1 & 1.0000000 & 1.0000000 & complete & positive & 1 & 0.7333333\\\\\n",
       "\t2 & R-HSA-1059683 & Interleukin-6 signaling & Factor3 & 0.6165494 & 0.7286493 & complete & all      & 1 & 0.7333333\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | pathway_name &lt;chr&gt; | variable &lt;fct&gt; | p &lt;dbl&gt; | padj &lt;dbl&gt; | view &lt;chr&gt; | enrichment &lt;chr&gt; | global_FDR &lt;dbl&gt; | coverage &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | R-HSA-1059683 | Interleukin-6 signaling | Factor1 | 1.0000000 | 1.0000000 | complete | positive | 1 | 0.7333333 |\n",
       "| 2 | R-HSA-1059683 | Interleukin-6 signaling | Factor3 | 0.6165494 | 0.7286493 | complete | all      | 1 | 0.7333333 |\n",
       "\n"
      ],
      "text/plain": [
       "  ID            pathway_name            variable p         padj      view    \n",
       "1 R-HSA-1059683 Interleukin-6 signaling Factor1  1.0000000 1.0000000 complete\n",
       "2 R-HSA-1059683 Interleukin-6 signaling Factor3  0.6165494 0.7286493 complete\n",
       "  enrichment global_FDR coverage \n",
       "1 positive   1          0.7333333\n",
       "2 all        1          0.7333333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(enrichment_result,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "52240082-f9a6-4f30-9369-305fe4e42216",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter down top factor genes out of feature weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "118e7e91-7958-48e7-8107-e90963416b52",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSelecting by value\n",
      "\u001b[1m\u001b[22mSelecting by value\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>type</th><th scope=col>variable_name</th><th scope=col>view</th><th scope=col>gene</th><th scope=col>variable</th><th scope=col>factor_value</th><th scope=col>factor_value_direction</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CD4.TCM</td><td>CD4.TCM__ERP29</td><td>CD4.TCM</td><td>ERP29</td><td>Factor1</td><td>0.548838</td><td>positive</td></tr>\n",
       "\t<tr><td>CD4.TCM</td><td>CD4.TCM__EDF1 </td><td>CD4.TCM</td><td>EDF1 </td><td>Factor1</td><td>0.534674</td><td>positive</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " type & variable\\_name & view & gene & variable & factor\\_value & factor\\_value\\_direction\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <fct> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t CD4.TCM & CD4.TCM\\_\\_ERP29 & CD4.TCM & ERP29 & Factor1 & 0.548838 & positive\\\\\n",
       "\t CD4.TCM & CD4.TCM\\_\\_EDF1  & CD4.TCM & EDF1  & Factor1 & 0.534674 & positive\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 7\n",
       "\n",
       "| type &lt;chr&gt; | variable_name &lt;chr&gt; | view &lt;chr&gt; | gene &lt;chr&gt; | variable &lt;fct&gt; | factor_value &lt;dbl&gt; | factor_value_direction &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| CD4.TCM | CD4.TCM__ERP29 | CD4.TCM | ERP29 | Factor1 | 0.548838 | positive |\n",
       "| CD4.TCM | CD4.TCM__EDF1  | CD4.TCM | EDF1  | Factor1 | 0.534674 | positive |\n",
       "\n"
      ],
      "text/plain": [
       "  type    variable_name  view    gene  variable factor_value\n",
       "1 CD4.TCM CD4.TCM__ERP29 CD4.TCM ERP29 Factor1  0.548838    \n",
       "2 CD4.TCM CD4.TCM__EDF1  CD4.TCM EDF1  Factor1  0.534674    \n",
       "  factor_value_direction\n",
       "1 positive              \n",
       "2 positive              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Get involved genes\n",
    "\n",
    "## Define gene-set to merge\n",
    "geneset_oi_pos_per_factor = feature_weights_long %>% group_by(variable) %>% arrange( desc(value),  .by_group = TRUE)  %>% top_frac(as.numeric(top_var_thres))\n",
    "geneset_oi_pos_per_factor$direction = 'positive'\n",
    "geneset_oi_neg_per_factor = feature_weights_long %>% group_by(variable) %>% arrange(desc(value),  .by_group = TRUE)  %>% top_frac(-as.numeric(top_var_thres))\n",
    "geneset_oi_neg_per_factor$direction = 'negative'\n",
    "geneset_oi = rbind(geneset_oi_pos_per_factor, geneset_oi_neg_per_factor)\n",
    "\n",
    "colnames(geneset_oi) = c('type', 'variable_name','view',  'gene', 'variable', 'factor_value', 'factor_value_direction')\n",
    "head(geneset_oi,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "94b59daf-1d0d-4e9e-a73d-57c9adc85806",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge genes belongig to pathway to enriched pathway sets\n",
    "\n",
    "pathways_vis_genes = merge(pathways_selection, pathways) %>% mutate(pvalue=p, view_text=view, cluster=view, name=pathway_name) %>% dplyr::select(-view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "650d5395-8fd7-4292-81be-0b20b80f2f13",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add the feature weights to the corresponding genes from geneset_oi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "54521e2f-8f1e-4323-b191-06b708358ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_vis_genes = merge(pathways_vis_genes, geneset_oi, by.x = c('gene', 'variable'), by.y = c('gene','variable' ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a7e99cc4-7b0d-4a22-9d77-6aa9efb7896f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene', 'variable', 'pathway_name', 'ID', 'name', 'cluster'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "### Summarise to get max/ mean factor value of each gene per pathway (remove cell-cluster/view/type dimension)  --> tBD max or mean (for max --> absolute value?)\n",
    "\n",
    "pathways_vis_genes_summarized = pathways_vis_genes %>% group_by(gene ,variable,  pathway_name, ID, name, cluster ,view) %>% summarise(factor_value = mean(factor_value), enrichment_type = paste0(unique(enrichment), collapse = '&'), pvalue = min(pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "2e5894a5-a557-49cf-9583-92697e679b73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>variable</th><th scope=col>pathway_name</th><th scope=col>ID</th><th scope=col>name</th><th scope=col>cluster</th><th scope=col>view</th><th scope=col>factor_value</th><th scope=col>enrichment_type</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>C1R</td><td>Factor3</td><td>Complement cascade              </td><td>R-HSA-166658</td><td>Complement cascade              </td><td>complete</td><td>proteomic</td><td>0.2727024</td><td>positive</td><td>4.589877e-07</td></tr>\n",
       "\t<tr><td>C1R</td><td>Factor3</td><td>Regulation of Complement cascade</td><td>R-HSA-977606</td><td>Regulation of Complement cascade</td><td>complete</td><td>proteomic</td><td>0.2727024</td><td>positive</td><td>1.137923e-06</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " gene & variable & pathway\\_name & ID & name & cluster & view & factor\\_value & enrichment\\_type & pvalue\\\\\n",
       " <chr> & <fct> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t C1R & Factor3 & Complement cascade               & R-HSA-166658 & Complement cascade               & complete & proteomic & 0.2727024 & positive & 4.589877e-07\\\\\n",
       "\t C1R & Factor3 & Regulation of Complement cascade & R-HSA-977606 & Regulation of Complement cascade & complete & proteomic & 0.2727024 & positive & 1.137923e-06\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 10\n",
       "\n",
       "| gene &lt;chr&gt; | variable &lt;fct&gt; | pathway_name &lt;chr&gt; | ID &lt;chr&gt; | name &lt;chr&gt; | cluster &lt;chr&gt; | view &lt;chr&gt; | factor_value &lt;dbl&gt; | enrichment_type &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| C1R | Factor3 | Complement cascade               | R-HSA-166658 | Complement cascade               | complete | proteomic | 0.2727024 | positive | 4.589877e-07 |\n",
       "| C1R | Factor3 | Regulation of Complement cascade | R-HSA-977606 | Regulation of Complement cascade | complete | proteomic | 0.2727024 | positive | 1.137923e-06 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene variable pathway_name                     ID          \n",
       "1 C1R  Factor3  Complement cascade               R-HSA-166658\n",
       "2 C1R  Factor3  Regulation of Complement cascade R-HSA-977606\n",
       "  name                             cluster  view      factor_value\n",
       "1 Complement cascade               complete proteomic 0.2727024   \n",
       "2 Regulation of Complement cascade complete proteomic 0.2727024   \n",
       "  enrichment_type pvalue      \n",
       "1 positive        4.589877e-07\n",
       "2 positive        1.137923e-06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(pathways_vis_genes_summarized,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "3f7fad8f-f5a4-4e8c-a59c-807e28e38565",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot the pathways with summarized gene values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "893814e5-9298-46d6-988f-61eeec026525",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_sum_plot = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "c9ac9527-78c9-48b2-8ef3-fd8ac1ee2ad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in unique(pathways_vis_genes_summarized$variable)){\n",
    "\n",
    "    xlabel = xlab('Gene') \n",
    "    ylabel = ylab('Pathway')\n",
    "    \n",
    "    pathways_sum_plot[[i]] = ggplot(pathways_vis_genes_summarized[pathways_vis_genes_summarized$variable == i,], aes(gene,  pathway_name, fill= factor_value)) + \n",
    "        plot_config_heatmap + \n",
    "        geom_tile() + \n",
    "        scale_fill_gradient2(low = \"#1D2ED8\", mid = \"white\", high = \"#D8911D\", midpoint = 0)  + \n",
    "        scale_x_discrete(position = \"top\") +\n",
    "        theme(axis.text.x = element_text(angle = 90), axis.title.x = element_blank(), axis.text.y = element_text(hjust = 0, vjust = 0.5)) +\n",
    "        xlabel +\n",
    "        ylabel +  scale_y_discrete(labels = label_wrap(50))+\n",
    "        ggtitle(paste0(i, ' values of pathway genes (top ', top_var_thres  * 2 *100, '% of features)'))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "7a75f244-225e-42fb-8458-004fb24e6f75",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32my\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32my\u001b[39m, which will replace the existing scale.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAACWAAAASwCAIAAADwxubWAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdeYCV8/7A8e+0b9OqfZ8WlbRTtE1kaw8VUoqKUFG4dtm3i6KEkC37lVtcrqVF\niCRS2pO0a9G0Teuc3x9TqWYmy03j1/N6/XXOec55zuc85znXnd7neU5cLBYLAAAAAAAAQDRk\nyewBAAAAAAAAgCNHIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQA\nAAAAAIAIEQgBAAAAAAAgQgRCAACA32fhM10TrxqblDlPvnv5+H9eeV7b01ok3vD+jsO+9pnD\nzk684cPDv97/f1Jmjzivy7DZKZk9x19n26K37+xzbquWLc66b+rByzbNfPGmHmefdUrieU/N\ny4zZfr8fRnU/+8Gvd2b2GAAA8P+WQAgAABxOU+47K/EAZw+b+SdWs+P9GxI7/alHhhDCjlVf\nvnLXxW1OSez61MI/uYr0bF785aRvV2ROk/j52YvbDFneoPc/bht8QZ1s/+va0mzepIWfT5q5\n+iiuYr/XipEDblrRuG2NLCGEsOq1yxN7vbj8MK4+tmnu2CHX9urUpk3niwcN+eDHbfuW/PRi\nrwM+N20e/HrvY9Z8MvTyTu3O6XXPf5fv3nf/lAVPdzvvsVm7/vAIcx/p3Oml7WdeeeNtN3ao\nfOCilCmDW/X+b4Fzrrpl8FUtS/2Jl/e/fWz/kIRWTbfc0/+xw/n5BgCASPmf/64EAADYz7q5\nkydtuOTtIR0L7LkhR5lyf2I1KatmTpqS/OcO1pt8VbWun5944ek5fxj15Q+b/9Qq/n5mTp26\ns8Vz17RreVjW9r9s3qNZyldD75/S4v4x5VOvbls6fdK0UsmHb/0fXlFn4Mr2Pdqcl1h8+8yX\n7+pw/Bt3fTFx4HHZQwhbl0yb9PNJrz/epWjqXbOVTEi9sOKpzm1HVn7orvaz7u3cJmXq9JuP\njQsh9uPwPnduu+yrmn/4b/qdX0/9tkSrR3u1apx22cqpU5dV79i/6xmV/uTrO5L7VfELL29z\nzeUPf9L/8Wb+YQMAAP44/z8aAAA43ApWaZyYeMyBt339cPtBY5NCyJK7SNmqjTpd3b9NhZx7\nlyX/8O6Tw16dOPuXnBUbXzBoYPvK0+9rf9+UsObbfonfFgih4iUvjOpWLsTWfvXsIyPe+XrZ\n9vwVG3ceMODcGvlCCOHTO069O/eN95f4+OG3pi8tdN7Tz/aofctXi4oUyT5r8Ni7Z2Q04rYP\nbmo7NN/tY284OeueW354utvFX5z+wtPdyh1q1L1mDuvUb/klH9x7Zo4QQghJYwe1f+3Yp0f3\nqRxCCClrpr7w6FPvTFu2s3BC/XZXDuySOmnY8O3ooSPGfb00uUDlk88dMKB9pVzpDLbonWGP\njJ6wYH224rXO6D3o0qYlsobFz3Xv+dB/k3bH3ZyY+EhoctMHd52WY7+HfHrHqXfnvuHeYpOe\neH/Gkq1FGnS99vrONfKGEML6Mf3OHjozhLjs8SUr1jqtz6AeDQrHhTAl7eYNIYSw/Yd37n/y\nzS8WbchX9fTLb+x70jFxYfO7g9q+XGPk6EsqhxAWPtW118u5Lxn9dLfSIYSZw86+enW//9zZ\nYnP6z7L2rQGdXqky/I0ra+yZM2Xqgx1uWN377X+2jT/wNe9c+sGQe0aO/ylWslqTrucVeeXa\nKWe+8fi5qaEsedE7Ix59ZcK8DTlL1WjebWDfFiWzhRA2jx3U5rWq9/XPNu7ZD79bvrtEw/P/\nce3Zlfduzo0zXx/2+L8+W7w1X7lap/cZdHGDwnEhhLBlwb+HPfr6lEUbc5c/oW3fqy+oddAY\nIcQmPPvcylaPtcsXQgjJ799w3ojZYfXqrokf5Q6h/sB/P9SuQAY7Yeo8d1+WMvb5Cd8v312q\n+SU3DGxVMcfB6w8N7/l6RsGCqWfyad26bvLxNR4Y8dnAYYl7FuevdFJiYpkDH5L03tuTmgx6\np1eHvLuyf1j4ln8vvvm6hLD0mb4P5rtjapdiaXegfTbOem3Io29O+XFTnjIN2l8xsHv9wiF8\n+3D7fk9OTlkzrV/iN/nztX7gnWtP3Hf/8bcl3vDit2HN8q6JH+YqfcHTo/tUzmhP/t37VZ2x\nGX1MUrfYPVfE3hz+n1krK/Ud80jHAn/wjcvdqlPrDReO+u/QZq2zH2I7AAAA6XKKUQAA4EhI\naHPd4MGDBw++6crO9ba/0b3Ouc+vTl2wbfrdLep2fmpltQ6XXnpunU2jOg16P1TtcF2HqiF/\no4sHDx48ePDlzY4JYfuU65o0uf7TAmf1vvKik3a8dtGJrYbOj4UQwtrZkz4YfE6nN7I26zHo\nlr6JxULIX6TIbwaDXPUrZf/woWfG7ztf6OyXHn5pVZm65Q416n6SFk6ZNHPVvlNy7lwxY9Le\noxVXv31xg1ZDllftcHn/S04vPv36E08e/PWOEMLSkV2aXj2x0GkX9+/TpvrGF865/F9pf/Jv\n2XOdGpw7cl3trpdf1qH0d3edcuJlH2wKoVjilYMvqJUtW50LBg8ePPi8Wgd90XPt7EkfDO7Q\n9vnk+p0u6dF4x7+6N+70wooQQgj5Tkzdgrde26tlqVl3Njnpxqk7Qkhv84YQwsSBHYf+UrNd\nrz5tCk+5qUXHoUtCCPnK5VrxyisfrQkhhDUfjn5jypQXX5uwOYQQfnz/xTFJ+RJyZPgsxzSq\numvM/U9/vncj7fhwxP3vZzu24cFZbuvHVzZu/cgPVbr07nZa0S8uP+Oy0ZOmL9ue+pAZ97U8\n8YoPcjfr0e+K8+pufvHc2ue98nMIIexaMWPSvwZ1uHxygcSul3U7YcOozs2v+mhrCCGEzROv\nadz8lq9LntG7/6UdKi955JT6fT/aHELYOKZP4+5v7mxyYb++59aLe793jydXptn8s8aP/7lu\no0apOThHnQuuOr1MKHFK/8GDBw8e3L1Bnox3wl0rZkz618BWPd7L1fyC3hfUWv14h5Mveyed\nA+ny762DIYQQV7BggZC8dWts3y2LX+3b9ozWnS4e+OC/F+49bjEpKSln4cJ5QwjZihTJn5SU\nFMKqly6/Y9etj19UIu0T7JUy+6HTGvYaG2vS88reLfNNGHBy01u+2hFChTbX3diucih4cq/B\ngwdf3+6AM4zW7Dy4f8tSoVTLqwYPHty/ZYmM9+Tfv19l/DFJ3WLtL/2kaKs+/7j1ohPz/Ik3\nLvvJJ5+w6aOP0vyQIgAA8HvEAAAADp9xF+UNBSqf3HyvPqNXpr3T6uFNQ537FsZisdiy4S2y\n5W87es2+ZZs3b47FYsmjWofSAybvu3XpsMTshbuO27zn6k+PNc9e4Ly3t8RisTFdsobyAybv\nTPMkM287LlT6xzcZDZo0um2uAheNS069Nv3GY0PtOxccctTYNzdVCs1HrInFYrHJA0qH1qOS\n995rzYhT9zzX7s/6lc7V5rm1e5fsmHhl2ewdRyfHYmO65ix2+fh9K96yeXPKQc+1a+JlxbPU\nv3funqvbJ1xaKq7OXbNjsVgsNq5rzpwXjUv3lYzpkjUU7jpmy56r8++qm6XsVZ8evPLYzo/7\nlsxz4dspsVjazRubPKB4KN7ng617rk67vkJoOnx1LBZL+aDPMbk7v749Fkt+vXPORpdeWiv/\nxf/ZFYtterZVtgrXfJFmmP2f5Zfn2+Yuesl7O2KxWCy26fXOefN2fG3jwQ9YMqRJ1uKXfLht\nz9XvbqsZQsNHlsZisdjKJ0/LXuWaafve2tVPJGapcsuMWCz2y4hTQ5aTH162Z8Hm59pkLdr3\n41gsFpt79/FZTvzn4r2vPmXu4OPjmg5dFotNGVA6S+fX961rx+bN2w8eJWVMl+xxnV/fte+G\nxQ823H+vyHgn/GXEqSGUHzBlz+p3fXpVhSx17pqXZuvsb/MnV1WOK9x1zIbUq3PuOb1xn3ue\nfWvcW8/ecV71fLmOu/azrbFYLLbzvUsKVBn01fbYrvn31c/TdnTSmlfOKZf42KJfpj7Zv0vb\nDj1uffuHHWnWvfHVc+NznzFy9Z6rG145J3/Olk+sjsVisU0jzwjlB01Jd6QF99YP9e9dnHol\n4z35QIfarzL8mKRusdp3zt931z/1xv27W+7Q9vl0XwsAAHBoTjEKAAAcbhVaXTd4728QxicU\nTL2waf7YkU+89cWCFWs37UhZPz/L4mMXhlBp96cTJqckPnvOryckzZs3bzqrnP7lV7ub9z5j\n76KybdvW7vf41Hmhfd0QQtZGTRr94b9t8nfo2ubS3qPfHdHmnNyxaaNfntfwiq6VDzXq71vt\nvI8/Xp5125NdWr4Yi8VCLBbbvmzTzqxzFoVQr2nTzddf06v8oAvOanFSzZJ50r7OJV9+ubpS\nm1bH7rmaI7Ht6XlHTp22PVRPc4rTgzU544w8ey5WOevMSjdPmroqNC4Zdq2a8tLjz384a8mq\nX5J3Jy/dvLXYwhUhlE5/HbUbNcq952K5cuXCv1atCqFYXOPEZruunvhl6LRr4iflWrzbdcs7\nPSZOD6evnfhprsTL64cQQobPUrBTr3P7X/jsuKFnnp13/WvP/jvnuW+2T3Naz2+mTc/SvP8p\ne1/h8aefVvL2z0MIIeye/PHEXRtX/ePMb/Z+s/WX+VkXHTNnR6gVQghlGjXa+0LylitXZM3K\nVbtDWD/+45nZVrzS67T39mz+2OpVsZ/nzA2hftOmhbvfdv4dWy5pc2rjOmXj09nNdiUn78xe\nMFfWNAv2OOROGPKfesbevTBr47NOyzt06tfbQ9UM3rldPzx/fufHdnV/a0iHPR+UigPGfJIn\nT5YQQmjTsXXFNpW7XzPiis8Hls92xnX3V2pxevWPiiatLHbbf06deHXDFf0+PG9qj6oPxt/9\nQOelj/Y869aPv7+37gFTz/nyy00Nrmm19wykBdq2bbZ9wNQZ4dLTMnptaWW8Jx/3x/arDBVv\n0qTK3str/tQblytXrrB16x98WgAAIAS/QQgAABx+6fwG4eLh7U64JbnLTX17dipVMG/2L+8+\n/Zbk5BDC9i1bduU8psBvBbAdSUlbcxWM//XMofnz5w9JSXtO4pgrX74/8adNnnYXduh9wehx\nm885d8boV39qfP35FQ856u+zadOmUOLMK26+YP9akqtsuRDiL/339NJPjXz1+UFD+yyIq9v9\noWcf61LlgKyTlJQU4uP3S2j58+dPWZy0KYTf2j7Z4uN//T3D+Pj4sGHDhhDiJ1x98ln/Of7a\nay9oVeGY/LlWvXzpl1MP8VKyZ9+3fePi4kJKSkoIIeRJTDxh9YgJs+fsnritxZA6DTc3X3nv\nhJ+mr5u0pcntTbOFEDZn/Cy5W/e6IP70Z99Yf/Zpo5/9sHDX8WeleSE7Nm5MzlM0/tcTb/66\nAbZv2rQz+/Gdb7qpSdx+DyhSLWuacffMG0vd/PkbXnzzvh8+DCHclrXE8SEUPOf56YWefeKl\nN287u//MrVXPvvuZkb1r5w77y16sWMEdK9dtDiFfetvn0Dvhge9cfHx8bN6GjSEUTWdFu398\ntespvWc0GTXx6Xb7PiY58+T59R7Fzjijbuwf384IoXyIq3rpu/Pbzpm7Nr7KcQWn9Koz+9Jx\nz1X4+Lxxx/df1bdj0ViRyc/0GvP9vXVr7f8MB+9JefLnz7Y5KWl3CBnWzzQy3pMP8Y7/Ifny\n/bqd/9Qbt3vduqQcxQ71Q4wAAEBGBEIAAOCv9/N7r00sc/XspwZVDyGEsHTGP3aG4iGEkKdS\npRLJn836IbRLOOABWbIc8IPpORISSm/9Yv7yvUcppcydOz8ktD7wMX9U7rO6np2rw+gx646Z\n8urqFrd1KX3oUQ+QJ0+ekJycHEJql/v555/3LKiYkBC3Yn3W+omJaY6VC3mObXPVP9tcFcKW\n2f9s1+CiG07p8Oa5+xezhISEsHj+/J2hYmqE2jB37qr4hIRj0qwojV3z5y8OIXVz7Jw3b3Fc\nxYoVQpjyxmsr2g+ff2enbCGEsPu9F9fuvf9Bm/eQSiYmVu0/4fXXUhY2u6xxlhzbEk/q8tao\nseuX1D07MX8IIUzN8FlCyNqsd48KJzz70sTloz6v2H14s7R5KkeFCiWTvl2wJpyZmtK2LVy4\nbM+iPAkJxXd8vK1o88SacWkel4FSCQk51y5LOTYxsWTapyp7ymV3n3LZ3WH7T89d0OCSAc+0\nn3jlgXGpfoMGWe6aMSOExqnXD9xMv7ETrpo/f1NokvquJ82btzpfxYrp1sElr1/Y4qIpDUZO\nfKVbxQz/Il+zZk3Ideze6Jslb6nj6pcKW8dffuU33V4feVzWjZ8lpRQuXCiEEFe4cKFfI+Ve\nCQkJ4ZP580NIPcQzLJw7d1e5UxN+fx0Mh9qTx//u/SrDj0kaf+qNmzljRsqJ/Rr+kVcFAADs\n8fv/KAQAAPiz8sTHZ109Z/a6EELY+dOr1w2ZundJk4v7VP9u6NVPzk49UWDSN0+99lUIIUex\nYgXWr1ixfe/dGl3Us/rUYbe8szIlhLB9weO3j1p/Ss8u5f63qXKcdmGngu+N6vvo6+vPuLBz\n0d8adX+Va9bMNe3dcatSQgi7Vo69/9mZexYU7dz3nDxjbun31o87Um/YOPuNh177PoQt4598\nZMKy1BvzFitZIGuIOzh7FejYs2Ps1bsfmrU1hBDWjr/l4Y/K9Ox5yu95KV+NuP2dVSkhhK0z\nH77vjZROF7XPHUJ8fPyOBbMX7AwhhKSv7rz95fX7XvlBm/eQaiYmHvPFkCFTT2iRmCeEwomJ\n1Sc/8tj0ComJ5UMIh3qWEEKoe8kltT+79/zHvq3b8+I66a39pAsvrDR5+O3j14UQwpbvHnzw\nrX2njGzaq2/NhY/1v3fqhlgIIYSdKyY99sT4jYeaNVfry3qW+/yevk/M2ZJ6w7Yf/nPf81+G\nEL56/t6xi1JXnbNIycI5U+LSbP9QsHX7pis//Gje3utFixULq1asiO25euidcPcHQ+74bEMI\nIWz4/PaH3y/V/aIWacZLWfpm9xbdPqv/1MRXL0o4oA5+//qQsYu3pV7ePGvYdY/PKdm5U9P9\n75H82c19J3UceWPdbCEUrlGjxLTJnyaH2NJJk5ced1zVg56n0gU9my195vbnftgZQti97M1b\nHp9Rp2f3dLd/hjLck//AfpXhxySNP/PG/fjRR4vrt2/7R89sCgAAhCAQAgAAR0K+zoPvOf6j\nC6pUOeGE48rWuWFr3YZ7D47KWvfmsS92XHpjvdIV69SvWarGpdNzFg8hhObdLqvwXu9qDZom\nJvZ88aeQtc71Lz1ce0KXquVr1a9RrvbNy9s/M7J3BmVg/Zv9EhMTE3s+tzgsf7lXYmJi4uVv\nrkn3nllP6dql6IQ3/r31rAvPKfybo+4vf+fb72/4Te/qVeqfUKvCSffnqFdz75LCXZ557591\nv+xRvWTleifUrnhMlfP/FUoXDyFn2SJL72paqlSNExrVq1D5yjntR9zV/uATbhY+/7EXeiQ/\neEK5Y+vVqVipzVtl7xh9R9P0nv5guZo02jbo+Aq161Uvf8Kdv5w38qFzC4UQThjw4EVbHmxQ\nofaJdRMqth5XrmnlfQ84aPMeUlzjxGa7k5LK7zm4q0piYsGkpFyJiXuOTjvEs4QQQsJFvRLX\nrVrb5JKLDq5YqbKdcOOztxR99axKFevUO7Ziq48qNi8bsmXLFkIIWevcPO71S3Y81rxs2eNO\naFC9VImT71hUqGzudFezV+7Eh98bdeZPN9UtXb72ifWqFCvTasTGEqVDCKVLbHumdfniVeuf\n1KBKuc4f1Xr4nz3THuBX5qIB5yx7/tmvUvasrFXPHjmf71i1YbPExEFjkw69E5ZsUeu7rlUq\n161bqWKL53JdPeqOFmnOp5r8cr8LXl6cJ/eqURe3TNzjqjFJIYRQvMSO0e0rlji2/om1K5Rp\ncPuac5557+HT9nupO6bddum7rZ+8/YTUlTa5+oGG751dvU6N+oO3XnNntzSHuFbs+/STZy7q\nf3zZ4+rXKl+tx7QTh75wTc0/+Pd/RnvyH9ivMv6YpPHH37jvn39hbqt+F/9vhxEDAEBkxcVi\nsd++FwAAwO+zbu7kmdsSGtcpnf3gJbGtq+bOXpyUo2T1GhV2L5w8M1a9efV9Z8/cvmb+zEWb\n4hOOq1os194Du3Yl/TRn/k/rt+SscOIJ5fOEEMLujUtmzl6+I75CzRql8uy529rZk+ZkPb7p\nsYX3PdGOZd98vvCAky7mr3xSvTLp/o5fbM33n3y/Jr7SSfXK7rc8g1E3L/5yWlLZxnVKpb62\n2LY1C+b8uK1Q5WoV8q2d8dnS+AYNE/b+qFps28/zZy3akLPUsdXKF/x1U6RsWTFvzrLt+ctX\nrVI8T0bnzdy+eu53i37JWqJarYRC+44yWzfnk+/jjmtWrUja+799XrYL872z+cmmKxfMXbK1\ncPU6FQv8moJ2blg8e96qnQUSah5bYM03X6wu0rBB+dxpN+/OhZ9/k1yp2fHFUx+5c8W3ny0t\n0KBhxdTXs3nxl9OW5K3erGbxLCGEnStmfDZ/V6VG9cvm+u1nCWH72O4lO697aOW7PQtl8IJD\nCFtXzJq5LFai6rFlvryiSNtfnt785rk59i7bvXHJ93NW7CyYUL1y8TypZ8nctWLGp0vj6zdM\n2HP2yw0LPv92e6XmNYvv3aY71y2atWBNXLEq1SoW2bdDhdi2NQtm/7A5d5kqVUvHZ3C+zYWP\nNDl5Qp/ZY7vv2Td3rFs0Z+GKDcn5qzSuXSp7SH8n3PBEy0L/bPDVwruqLp0/b8WukjWPL5M3\nnbd398rvJs9bf+BtBavs+7DEtq//Yc7CjXlKV6pYOv9Bn59f5n+2MEfdEyr8+kOFseTVc+as\nypNwXIWCGZypNLZl+fezlmzOXaZGzXL59+4Su1d+N3lJ3nqNKuVP+4Dkn77+ck3RRvXL7Rel\nM9iTf+d+lSejj8nB7+C+9f7eN27jv8+veutxH02/ueYfOnEqAACwh0AIAADw/9qeQPj0mZk9\nSLpWPt0y4Ybyby155qw8Gdzj5/GvfVv9nNNLZgth03f3tG18d95Hf3q3Zzol9MjYtXLGZ+tK\nNq1Z7A8ccLc3EN7X4K+biwOsm/vJkrwND+j6AADAH5DhT6IDAADA/2D2sE59npj89ZqWw5/N\nsA6GEOLzLx/SpMxl+cvGr18wN/m4y18dcVGm1cEQQraStZuXzMTn53cpUq1ZZu4lAADw/54j\nCAEAAAAAACBC/uCPlAMAAAAAAAD/nwmEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhAC\nAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEABy9UpLm\nvPf4NeecntlzAAAAAAD8jQiEAByFtvw4adSt3RuXL1mj450fb6uR2eMAAIc07+WrH5ywObOn\nAABCCGHL4k/efOapF8Z9vWLHfrdunfrcA//5MbNmAuDwEwgBOIpsXzX1tfv6nFalZOWzrnhg\nTPyVH/24dsU37w7J7LEAgENa+fmL736/LbOnAADCxvFXnVizeee+/S5p16ByrW7Pzd6+Z4FA\nCHC0yZbZAwDAYfDL7HEvPPXM0y+9Oz9b7Q49rx3z5OmzL24wr+HJ5fNl9mQAQKrVX7354bz0\nK+Cc+dtDtSM8DgCQ1rzh147M1/+TVXc2jV/+/sN9L+nVvF3SR28PqJ07swcD4LATCAE4Goy+\nvN3Auafc8PjUAR3rFs0eQlg7J7NHAgAOMOf5y7oNX5fR0ubtjuQsAEB6dn791ey2137RtFi2\nEMqfedM7U4+/vGXnU9ulfDT26jKZPRsAh5lTjAJwNDi+5WkJyRMfuXbAdfe++PkypygDgL+f\n4+rWLdXj/W2xdEy4okhmTwcAhBCXPXvWlJSUvVezlG43YuKY81bc2LLdkBnJmTkYAIefQAjA\n0aD5zR8sWPHDu7e32PXBTS0rlKrZ9rp3lmb2TADA/oqe37f5f594c0NmzwEAZCRb48SGX036\nfNevt8QVP2vYhLcvWHlDhzumZd5cAPwFBEIAjhZ5yyf2uP3FT39cMevlvtU2/rx509Ndarfq\nfdtT736b2ZMBACGEPO1uGd40fnUs7ZLa/d56qEOBIz8RAHCgUj2ub/XFqLEHfp+n2BmPThjb\np0q2XNn9WzLAUSQuFkvnrzMA+H9v55rp41545ulnXv7vz7/sXpvZ0wAAAAAA/F0IhAAc5ZKX\nTc9dpl5mTwEAkbfkvw9/UqRvtwa5Qwi7d2zZEZc3d/bMngkA2F/ytBdHrGs28IzyIYSwM3lL\nLEfeHFkzeygA/hoOCwfgaBFLXv3DwhWbd4cQYkmz3njohquuvvmh177ZoQ4CwN/B4nH3PPPF\nltTLkweWP+tJB/gDwN/Mli+euWfc4tTLa588q/zAyZk7DwB/oWyZPQAAHA5rP+x/asfHvtsS\n4uvcNPaxbFefdteCAqUKbl899MGR0z6a+2DTzJ4PAAAAAODvwhGEABwNZj8+6Om4bs9/OGHs\nzWVfPr/T8yX+OXfNqp+WrVv6Qa/YsAcyezoAAAAAgL8RRxACcDSYPn1u6xu+6t4yZ2hZdOqo\nuj/06Vs5dwghS7GWA7pWaZXZ0wEAAAAA/I04ghCAo0FSUv7ixXOGEEIoXrx4/kKF9v0HrnDh\nwhsybSwAYH+T+hWNi4uLi4trMXzdvsupEof5SUIA+BtYN7xF6n+bi/abtO9yqmOunJjZ0wFw\n+DiCEICjQSyW2RMAAIdWrduwUQ22ZbS0RKP4IzkMAJCO+JbXjxrVI6OluapXO4KzAPAXi4v5\nJ1UA/v8blpj93g2NqxQMIexc/u3UrRVSL6denbdgg2MSAAAAAAD2cAQhAEeD6md2PW3unssV\nKlTZf1GFCvUyYSAAII3t37z48I8Nr+9YNW7/W5M+e+LpTe0GnVkqs8YCAH61/P2HxsX3uqxx\ngf1vjM0fc+/UCoMurJszs8YC4LBzBCEAR4NYLMTF/fbdAIDMs3xYixYL7wzB7xQAACAASURB\nVJw1pEmOA29fOeLUZrNv/f6x5jnSfxwAcKRsn3hlzbtqfvLRZSUPvH3H5AE1b6s6cfwVvtAD\ncNTIktkDAMBh8Gzvcx75fK3vvADA39f6d9/6/tSOB9fBEELJjh1Lv/XWl5kwEgBwoC/feqtM\nx44l09yeo2nHU2a+9e4vmTASAH8RgRCAo0G1ssvubF7z1BvH/bQjs0cBANK1aNGihEqV0ltS\nolKlpEWLko/0QADAQbYuWrSpUqXi6S2qVClh0aJFR3ogAP46AiEAR4PGt30+870rcrx0zvEn\n9np+1qbMHgcASCN79uxJSUnpLdmWlLQ9e/ZsR3ogAOAg2bNn356UtD29RUlJSdmzZz/SAwHw\n1/EbhAAcRZK+eaJf92te33Lqlb1OOmbfl2DyXH99/8ycCgAIIWwf06XwzdWmzLi91kEpcP3L\nZyc8UO+Lb2+uljmDAQB7zbmz9knfXr/4X+cXOvD2XTNurdV4/j3rXu2QM3MGA+CwEwgBOKr8\nMu2R81sP/HhL4QK54vbeVmTt2nmZORMAEEIIOyZfVa3l2Fp3PTG0/+nlU/91MeWXb1+69bL+\nzxV4ZN5/e5bK5PkAgLD86ZbHXrv54qFP3H5hnUKp37vd/uMHjw647ObvO46f81BjxxACHDUE\nQgCOFruW//euiy+++4viPYe++HCP4/Jl9jwAwEG2zRjasfXA91fmKlouoULx7EnLfvhxRVKO\nWgNef3/IWSUyezgAIIQQVr7T74zzhs3aWaBUhYQyBXau/vGHn9ZsL9V66Hv/uqKmwwcBjiIC\nIQBHgy1zXrm2+xUjl9X+xzPP39aqnK80AsDfVGzT3P+8+NK7n89ZlrQ7d9FK9U7tfFHnhiX8\npxsA/kZ2rvziteffGP/NojXJWQuUqX5ym27dWx3rW7gARxmBEICjwfAWuYcec/cLT1zdqEjc\nb98bAAAAACDCBEIAjgZzps0u36BGnsweAwDI0M/Tx4yfvz2jpcXqn31KlRxHch4A4GA7Fox/\n6+ufM1qas+opHesVO5LzAPAXEggBOCrsmnZ/55drjnq4dYH9b43NeqzLiBLDhnfyFwwAZLZJ\nA0qcOnxtektSdu+ONX9szcQrjznSMwEA+1v3+KnF+09Kb0ksZXdK4SsmrB2WeIRHAuCvkiWz\nBwCAw2Ddq3c9UeDUUwocdHNczTYnTr35kUwZCQA4QPOhq3Yd7JdZb9zUtnKeuAK1TqzmTAAA\nkNmKXP5xmv9ab1s28bHeJxbNmrNMo7qlM3tAAA4fgRCAo8G0SZNqNGuaO+2Cis2aJqf77UcA\nIDPtXPnp8EtPrlzz/Oc2txo+bdE3D7QUCAHgb2bT7Ddva1+jcvMbpyQMHDt3wTuXVMnsiQA4\nfARCAI4GP/20u3Tp/OktKVOmzE9HehoAIGOxjbNfv6V99crNb/2q0nX/mTf/w4cvqlfEn6YA\n8Heyc/mkYb0bVa7V7fnt7R+f/sP0l647q0KuzB4KgMPKX2EAHA0KFdrx00/r01uyZMmSQkd6\nGgAgXTuWT3i0d8NKtXqM3nHuyG9++PrFQaeXz5nZQwEA+4klzXr9xrbVq5wy+Ntq//hg/rz3\nH+xeu5B/QwY4CvkfdwCOBo2bN570+KMzdxx8+89vDHk5JGbCQADAQRa9cG71KqfdNaPWzR8u\nmPvefRfUKhCX2SMBAAfYOvHmhpVq93gj1mnUjEVTnxt4Slnf5AE4asXFYrHMngEA/mdbJ115\nfMtXC3W75fYrWterUqbAztU/zv7kpftufeSzYx/79v1eFTJ7PgCIvIlXHtNi+C/ZcmbPmt7S\nZkOWf3BZkSM9EwCwv7XDEov2m5Q1R85s6X2N55i+Hyx7pNkRHwqAv4ZACMBRYsei1/t36Tvy\n6/Up+27KXbH97aOfu/akgpk4FgCQatWXr74/Z1tGS0s0Ov/Mag5SAIBMtX3u+698sSqjpbmq\nn3lewxJHch4A/kICIQBHky1Lvhg/6ZtFa5KzFihT/eRTE2sUyZbZIwEAAAAA/L0IhAAAAAAA\nABAhWTJ7AAAAAAAAAODIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAI\nEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAA\nAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJE\nIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAA\nAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCB\nEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAA\nAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRC\nAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAA\nACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgB\nAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAA\niBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQA\nAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAg\nQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAA\nAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAI\nEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAA\nAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJE\nIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAA\nAAAgQgRCAAAAAAAAiBCBEAAAAAAAACJEIAQAAAAAAIAIEQgBAAAAAAAgQgRCAAAAAAAAiBCB\nEAAAAAAAACIkW2YPAH87W6eOyOwRAIBD2b1jc2aPAAD8hl1b1mT2CADAbyt0xgOZPQKZwxGE\nAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAA\nAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhAC\nAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAA\nECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgA\nAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABA\nhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAA\nAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAAR\nIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAA\nAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESI\nQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAA\nAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECEC\nIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAA\nAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiE\nAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAA\nAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhAC\nAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAA\nECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgA\nAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABA\nhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAA\nAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAAR\nIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAA\nAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESI\nQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAA\nAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECEC\nIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAA\nAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiE\nAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAA\nAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhAC\nAAAAAABAhAiEAAAAAAAAECECIQAAAAAAAESIQAgAAAAAAAARIhACAAAAAABAhAiEAAAAAAAA\nECECIQAAAAAAAESIQAj/x96dx/lULg4cf8ZgzGTfx74v2VIioRmUJUmLdllKpVWbLq1ab7nt\nm26RaFO3X7ektFoiRSUiDJJ9zz6DYWZ+fyDCNKWU7nm/X/PHfL/f5zzn+Z7x38dzDgAAAAAA\nQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEA\nAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAA\nESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAA\nAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABE\niEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAA\nAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAh\nAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAA\nAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQI\nhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAA\nAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQ\nAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAA\nABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAI\nAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAA\nQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEA\nAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAA\nESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAA\nAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABE\niEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAA\nAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAh\nAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAA\nAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQI\nhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAA\nAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQ\nAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAA\nABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAI\nAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAA\nQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEA\nAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAA\nESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAA\nAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABE\niEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAA\nAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAh\nAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAA\nAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQI\nhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAA\nAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQ\nAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAA\nABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAI\nAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAA\nQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEA\nAAAAAABEiEAIAAAAAAAAESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAA\nESIQAgAAAAAAQIQIhAAAAAAAABAhAiEAAAAAAABEiEAIAAAAAAAAEZL7r14AAAAAAAAA/PWy\nMnesmTjgD582T6Hyhetd+IdP+3sIhAAAAAAAABCyMrevnfLvP3za+DLHCoQAAAAAAABwWIqJ\n+XvM+fsIhAAAAAAAABBiQogJf3zMO+zyoEAIAAAAAAAAux2KnHfYJUKBEAAAAAAAAEIIIcTk\nOgSTHoo5fxeBEAAAAAAAAHaygxAAAAAAAAAiIybmEDyD8LDrgwIhAAAAAAAA7GIHIQAAAAAA\nAERFTEQC4WH3UEQAAAAAAAD4a8TEHJKfA9k6983bu5/arv05vQd+sS679Sx9rVfr5J4vLtr1\ncsK9rZN3a3//5IP9lnYQAgAAAAAAQAghxPxpu/22T+hzYs9pZz1y/SlrX+3XvmtCyjvdSu43\naOmgy5+amfbtxoVpu95YM2vcjmYf3HNSbAghtnS1gz25QAgAAAAAAAA7/Um3GE0f9eyQItdP\nfbBbtRDaxn5V6v6hi7r1qfDzMYsGXT6s4aP9Vrbpt/e7xWsnJSf/zsLnFqMAAAAAAAAQQvjz\nbjG6NCUlpmHDnVsA4xs2rDFnztx9RiwadPmwox6/9eg8+7z/zZNntu/Q+aJ+gyevzjzYb2kH\nIQAAAAAAAISQK7biBe/v/caG6S+v//al3zpN4fpdCtW74KeXqQvH7T8mNTU1Pj5+14uEhITU\nzZv3/jhr0cBeQxs/9slRecLHe7/f/JZPhq7M2r550ZdD72590oqJU26pdzC7AQVCAAAAAAAA\nCCEzY9HLJ+/z3kE8lXDDty9v+Pbln17mS2xYeK9euFNiYuLaacu3hRAXQli2bFnpxMS9P54y\ncMDEFWWuazMmhHVzNnwf17V97sdG9W0aitdOSq4dQggnnVxsbtE73p1/S72DeRChQAgAAAAA\nAAAhhHDA24H+/kn3f6tY6xPr3/Tiy0vPu6js1s+GvrGpzaMN9v64es9X3mq7LYQQwtcPdxpY\n4qqbT6vxs+NTp342dUticomDW5BACAAAAAAAADv9SYEwVOs14JLhZ9Wt/njZLfM3NnlqXNu4\nEMKGEdd2GnfSyIc6FKzaNLnqzoE7hufJX65xi1rFQgjj7mp9x+iMsGPz4plzc7V67O3zCx3c\nggRCAAAAAAAACCEcmh2EB56z0IkPfvXDldNmrYqr0qBOqXwhhBASGnXvX3HX77sdc/3bg+Iq\n7Pz9yM639z8hKyZPgZKVqtcoWzD2YFckEAIAAAAAAEAIB/XEwV8xZ3ZiC1c+umnlvd7IU+ao\n5DL7DCpSo1mj3b+XODIp+cg/YEkCIQAAAAAAAOz0Z91i9C8lEAIAAAAAAEAI4c+8xehfSSAE\nAAAAAACAnewgBAAAAAAAgOiwgxAAAAAAAACiI+YQ7PY77PKgQAgAAAAAAAC7ucUoAAAAAAAA\nREck+qBACAAAAAAAALtEohAKhAAAAAAAABDCoXoGoUAIAAAAAAAAh6eYwy7mHQoCIQAAAAAA\nAOzkFqMAAAAAAAAQHYdiB+HhtytRIAQAAAAAAIAQDtXzAgVCAAAAAAAAOEwddjHvUBAIAQAA\nAAAAIIQQlVuM5vqrFwAAAAAAAAD8eewgBAAAAAAAgBBCTER2EAqEAAAAAAAAEEIIMYfgGYSH\nYs7fSSAEAAAAAACAnQ67mHco5PgMwszt2zP/jIUAAAAAAADAXysm5o//ySY6bp375u3dT23X\n/pzeA79Yl916lr7Wq3VyzxcX7X6dseTDAZee0a7dGZc9+Mmyg254OQbCtc+eWqHZBbcOHv39\n5qyDPQkAAAAAAAD8DcQcmp/9bJ/Q58SeY4uf2evi41Y83L7r0FUHWszSQZc/NTNt6lcL03a9\n8d39J3d+JabNpZe2DS+d2emRuQf5JXMMhIU7XH/D0RtH3NS2emLV5IvuGjZhUVpOhwAAAAAA\nAMDfT8yhsf+J0kc9O6TI9c8/2O20zte9MODkcU8PXbTfmEWDLh/W8NF+DfY8MXDi4H+vu+jJ\nZ3qdccZlzzzRbcXTg748uK+ZYyDMXemk655459tliycNuqz2sld6J1VOrNnmsvtfm7R828Gd\nEQAAAAAAAA5Pf9IOwqUpKTENG1YLIYQQ37BhjTlz9t0NuGjQ5cOOevzWo/PseWtrSsrioxo2\njAkhhJiGDRv8MGfO9oP6krlzHhJCCCGu9LHn/OPYc/7xyKL3B1zU485+5z57a9E6HXvecEuf\n7o2KR+JpjQAAAAAAAPwvyxWbePoLe7+xefbbm2b997dOU6D26flrdfrp5dYlk/Yfk5qaGh8f\nv+tFQkJC6ubNe3+ctWhgr6GNH/vkqDzh472Wk5oa4uPz7XwRn5CQXlv60AAAIABJREFUtXlz\nWgiFfuv6fn0gDCFt4advvPD88y+8MX51iRY97urRvuz3/33ypmZvTPlgxhPJ8TkfDwAAAAAA\nAIevzIzlb/XY980D3SD0l22a/dam2W/99DKuZJ38tU/bZ0xiYuLaacu3hRAXQli2bFnpxMS9\nP54ycMDEFWWuazMmhHVzNnwf17V97sdG9T0uMTHv8uUrQigbQli+bFl8YuJB1MHwawLh1qVf\nvDn0+eeHvDZ6Ufwxnbr1++9z551YvUBMCCGc1f2MvrVPGTntieTjDurkAAAAAAAAcNiIOdDt\nQA/FnMVan1j/phdfXnreRWW3fjb0jU1tHm2w98fVe77yVtudj/v7+uFOA0tcdfNpNUKIOeHE\nVt2eHZpyxc01Q8rQV6e3ubr5wS0px0C4ZtAFzW5b1bbL5UMe7npq/eI/H5+rWuuuHZcfXJoE\nAAAAAACAw8uheLLegeas1mvAJcPPqlv98bJb5m9s8tS4tnEhhA0jru007qSRD3UoWLVpctWd\nA3cMz5O/XOMWtYqFEEpccM8tL7RrXO21SlkLVlS74+Ozix7kgrKysn5xQPqy+SuLVimf7+Cm\nh7+htMkD/+olAAC/JCN9c86DAIC/1I7U1X/1EgCAnBVpO+CvXsLhJWvHtqXDz/jDp40rWadE\nmwNe6oz1P0ybtSquSoM6pXamuO3Lpn62ulTzBol7bdlbN+ez7+MaNqqYsHuVqYu//XZpKF+v\nfrkjDrZm5riDMG+ZKuV3/pa5ddOmrRm7c2KuuPwF43/9EwwBAAAAAACAPWILVz66aeW93shT\n5qjkMvsMKlKjWaO9X8ccUb5B0/K/78S5fsWY1eMf6tK4fOH4hIKFi/ykSp8JOR6YuWnJzK++\n+GLqvBWbd/y+Zf69rZ09ftysNb9i4Kb5X4ydtnzPtdrx0mkxpa8ae8gWttuGj66pXah4zaZJ\nyec+m3LIzwYAAAAAAHCYion5438OPzkHwjWvXXHawz92emjIVQ0LnfXU2JHD7jnnyAKJZ953\nzTG/cFDWmgmPnH90qRJ12l987XUXdzimfNn65/QftfiX72b6P2vy/e1b35tzTw0h5dlzW97w\n7p47ZsWUqpt0fPXCh25lO21+6/EnUnuMnP35uLHDL615qM8GAAAAAABwmDoEeTDm0DzX8HfJ\n8SahWeNHvHf07QtvOTs8+XSuLbWSOrRK6tC2/Mm1Xpm449Ia2Ryz+bO+Sa2eKXzja/PuaFcu\nLoQQsjbOevOxCYuzQvmYNTPHzYqt16Jq7ILvZi3eXvroRpWOCCFkbFw4Y+bSbQUr1atdJn7P\nVdqxflHKvJUZJWvUrlAoz0/vbvtxfsr8tbFlatUqmz/2QAtIW/jV5B9LNT26VNrClLkrc1dq\nULtk3K4Jl02bsLjAMU3Kpc+dkbI8tmLTBmXzhBCytqyYPWNBWv4KNaqXKbDXNUlbNXfOgo3x\nFY+sXip+75aavmbu9Hnr48pUr1GhcN6fz1wxZvnclKUZpWvWKlsgNoSw6fsvvl2RkZUxY+zY\nwiHkLdvw+OqZcz6btmx7CLnii5WvVr1SsXwxIYSwdfHXXy/aGtalTBg7Nn+IKVkn6cji9c/t\nf2NcpT3n3b523vSU1ZlFq9WrWSJuny9beuviOXNX5y5fq3qphGy77/7XedV3Yz+eODfEJy4Y\nN3ZrfIVGTarkz+5gAAAAAACA/3GHIub9/QLh2mXL4qpXLx7CtoIFN61blxlCrlAyKanqzTNS\nQjjugIf88PR1D81rNGD2fe3K7X4rpmDtM2+rHUIIYcJdrc9bes4Za0dPyFul8lGXDBnSvdSU\nx88/s9/H26vUKLBy9vJS5z7z9rPnVskd0qc/3rnjLRNy1axRZP0Pq8t1f3b4A+1Kh83j7zrl\nrAHfFT6ySr5VP6TWuX7Yy32b7rfFbtGLPVsOK9+txJT3fixVOm1eyo4m/xzx1vWNjghh84gb\nWt6Vq+tR8z+YW7h62ebXv/3o6Xm/G3zJ2X3e3Vyheokt8+emN7v91ZduOK5ACKvevbZt18Er\nE48sF7N0UVazu1954ZL68SFkLHijd+fLX1xRukaZrCWz19bsPey1u08stWvm2O7HLPhwTv4y\nsYunLyt39Zuj/9Wq8LIPH31panpm1vD+/T8OoejpT7zZO33kgP4jNoSQkbpyzsxVFS8Z8uaj\np5YL6yc+++yE9WHTiPv7f5k7xLa685Pbm310a8sby41Z8WRyCGHtuDvPPO+Bb+KqV869ePb6\nGlcOfevBk0vv/rIVLy47/f3VRUtvnT99Y9173vuwT6N8+/1d0g54nWe83v+xj5aG1R8/2n9O\nvrLnD3r50mq/9h8PAAAAAADA/5ZDcUfQw+8uoznvIMzKypUrVwghrlat8n3femfdmZ0KLvl4\nbErxpOLZHLFq1MgvM5o91aVyNp+HsPXz6SXHzVzQrEhMCCFj2u1drpua9Oac5zuVybUt5fGO\nx118ybEtPrm8+Ft33zg56a0FQ08uGELmms+eHb0yhNKrht1yx7KukxYPaJwvhB2LPxgy9ccQ\nDngPzrmfzL/4i4X/qB+fteLt7k3PuOhf7ab1PzImhBCWT/j+mknz3quXP4QQto2/usMNy68Y\ns+imhgVCSPvuwfbHnXtb45RHW3z/ZJ/H4/8xd/FNVWND2DL3jeELNoUQnznj/k49Pms/8vv7\nkornCukLXjj32PN6HzN3+GmFQwhh+aSlbSfNfadmQtj86RUNkm58uueUm2tePnzApJGn7Lhn\n7Eun7V7b9W+PvX7nb+kLhp3f+KJ/tF7wcsfS5/z73998cOxXfUZ+3HPXN9rryY1bxvS94O61\n547+4eETioRNX95+YvOuvZvNe+30nSPnfpl65xc/nFchT9aPr59bq8vNw3p8eOk+f5+Mafcf\n8Dq3unPsy/GNqr/RefjYvpWy/ZMBAAAAAABEwWEX8w6FHJ9BmLtIheqlE0IIofHltzX4oHP5\nUuVLV7/s61bXd81uo9mSJUtCfKVKJX9h0pZX3rqzDoYQvnhhyKxjr7qnU5lcIYS4mlf171Fk\n9JDXFoXMjIysrG2paTtCCCFX8Wa9zm4QQtiRkRGytqRuyQwhhNzl217SsWo25yjUue/19eND\nCDGlO915ZcPpQ1+csuuTfJ2uu7Hervtopo/69+AVjds0SZ02fvz48Z9+9WO94+osHPX+zBAy\nMjJCRtrmrSGEEOKrd+5xUskQwsRBA6fXbtMydvbE8ePHf/rFospNjt48atSXu2YueMa1V9VM\nCCGE/Cd0TC743Yzvsn3q4ra1C2Z+PfHTiQvKNaizfvz4mb9wtUIIIfPD54etatfnzhOKhBBC\ngWNvvuXM9P8MGZG66+PKXfqcVyFPCCGmWMdTjt8+Y8ac/WbI7joDAAAAAAAQws4nEB4Sf/UX\n21eOOwgLXzDs812/Vu7xzoxjPhn77foC9Vu3rZ/dBsIQHx8ftm7YsC2EuOxGlC9fbPfv6fPn\nL02oWfOnm5HmqlWrRhg2f36IP/POJ/57Ts8qZe5q2qrliSef1/38pom5Q5nuA/45qkvHcv+t\n2aJVy5M6XtC9c8NiMTuWT5uQsi6EEEJs6XotahULIYRKNWr89NTCSjVrxi2aPz8jHBNCCInl\ny//0tZfMmbMl18o37rh11E+ry5dUteC2EBpe/Vif8Rc1TRxaP7lVqzanX9j9lNr5Q9qcOUtj\nl7z/z1sn7fk2TRoWy9yZEUOJknuyaHx8fPqPaRkHuMRZC1+//PQrhy7IX71qmcJH5Fk/O2t5\nueXZXc1dls6fvy3xuJoFflpkrVqVsl6f/0MIdUMIoeSeE8fHx4e0tLR9J8j2OocKOZwaAAAA\nAACA/yU5BsL0BVOmhmrHVCoYG0LIXbJ+27Pr53BEtYYN82d9PmVq5nlNstmfuPOepTvlLVQo\nYeumTTt+WsqmTZtCoUKFQshbt9fr3120etbn40a/O/Tulv96fdB3I7skFmjW9735vZd/+9nY\nj98ZdG2Thz94a+agE6YM6/+vr0MIIcS3++eovk13z7Pbtk2b0vMnFo7d//QJCQkhV9Kdo4ed\nvN9aS7d74NMlty2cMn7Mh28O7N7g6Z6fTru/UUJCbMZRvUe+d1H+fYdvzeGy/GT7iL49hpd7\nPGXyxRVyhxC2DT8toUtWtjsNdylcqNDPvtGey/QrZX+dAQAAAAAA2OVQ7PY77HYQ5niL0c0f\n9ju+ctEilY49uXufAS+8O3nBxowcjsjT7opLq/4w8NZBi342Mmvjho0HGn50k2NjPx31we4t\nb0vefXdaocaNa4atGzemh5C3RO2kzlcOeOuBUzd99MHkrJC6cWNGiIlPbHDieTc89p9bmi/7\n4KMZIX+Hh8busqsOhhB++PCD3ffZ3Pz+qAl5jj32QGWzdMtWtVPfHfbm2r0XunFTCGHTxo1Z\nIVf+io3ad7/5uSHX1Jr1wSeLQu4WrU6IHfviS4v3mmLLxo079p325xISEjLT03dfjVULFqRV\nOK5FhZ2hbsP7707I3GtgSE9PP8AUBRo3qbX+o1GTdk2S9d27oxaWady43AGGZiO76wwAAAAA\nAMAuMTGH5Ocwk+MOwqKXvreu7ZTxY8aOHTv2jf7P3rwwNaFiw+bJyW173NA7qfSB5zz+3jcf\n+a7d1U2Splzbs91RZXKvW5Qy9eOXXj/ivrmD28fuO7pcz/t6P5Hco0PJe69OLrTo7XtvnXzc\nvVNOTQiLnji73aiaF3dOqlMmLHz3Xx8WOOmJxjFh7sOtu85offGpTWuW2Dr7tYGTEtveWO/A\nC88358HTz4+97dxqqRMe6ze8+FVjupY60LA61z51wxsderToknLjmQ2LpS345qNhLybcNe/p\nNl/2PfYfqef2aNuoasGNUwYPnVvjrFaVQoi9+OG7X06+NunUJf0ubJqYuXT6uFeHrLn0y9fP\n+6WteDXr1ctz94sPvlW0SeGEsg2bJp9Y+8anrr6jwjWN4+aPeuje/0v9qdJWqlevwIOvP/p6\ntZNKxpWqk3RkkT1z1L12QLfnOp/d6Yi7ezbMPeulO+5b0Wlw32a/4d9TdtcZAAAAAACA3Q67\nmHco5BgIQ4gtUPHYk7sfe3L3PiFj3fQ3H7v7zkffGPrV5PwdsguEIeSr33vUrNZvDxn2zrvP\njd6Uu3jFmg07vzTljDqxIYTiRya1iC+61+C4pgMmTKj18NPvPPtEeoHKZ78wuXfnGjEhVLz6\nrfdqPDfw9ZdHr8gqVv2MF7+4vGNiCIn9R7/234H/fmvIBz/mKXVkrxGf92xd8MBrqH/biDvz\nvP7qC2NWxda7+cNXr26eN4QQcpdpkNSkSoG9xhVs+eDn37R4ZuAbb/17fEzJ6ke3eeqL848J\nIbR68vN/v/LU8yOfe2dzfLkGt4x+plvT2BBC3FF9R3/baNBTL496/vHthasc1fKusT1aFDrA\nzIWrH59UuFRMCCGUuujZN9YMeHHwgPc35T/tiTd73/7BqEL3Pzf88Ul5E4/q+fYbq/s/lLnz\niY4JnR5+u//9z7380N0bdiTf+cntzUvVTTq+ROEQQgiFOj43+f1nHhr8n4ETs4rW6vXfr65q\nv7N4JlRslHR0+T3PeyxxZFLz+L3SYg7XOYQQX+GYpGMq5MvmjwkAAAAAABARMYcgEB6KOX+n\nmKycHn+XuXnxNxPGjh0zZszYseO/WZm/dtOk5OTk5Nbt2zYtf7juP5t9z1G1R3ZZ/MWNv+EW\nnLBb2uSBf/USAIBfkpG++a9eAgCQgx2pq//qJQAAOSvSdsBfvYTDS1bGthVvdPnDp81bvFax\n1nf/4dP+HjnuIPxxYMcKV00s0+LCy7r0HfZwUpMaRfP8GesCAAAAAACAP9uh2O132O0gzDEQ\nxtc/+aykVZ9OGvbI8plff/lVy5bJycnN6pdJyJXTgX+lfe+6CQAAAAAAADmKOQQx71DM+fvk\n2PkSWvR5fex3K9Yv/ey569qWXT/x2WvbVS1SombzC19I+TPWt8e8wRckXztiw68aW+HCQWOf\nPqfEQR0LAAAAAABAVMUcmp/Dy6/dCBhXomqdunXr1KlT58haZePXzfns42mH9k7y6e/3Sz7r\nyel73tj8w6RxU5dtP6jJftWx+50RAAAAAACAKIlEHvwVtxjdNvONAQPfGDt27MTvVmYWq3Fs\ni+TkawbemtyyWb3Sh3RhmSumj/t8y16b/qpd/PKYzuUL/YlnBAAAAAAAIFJiDsFT9g6/W4zm\nGAjTprz/wdJCyZc80i+5ZbN6peN/zTcYcUPyazXu7ZU5YuiY75ZmlEm6uN/1J1fOG0IIPwzu\n2uPFRSHkiitcpuoxHa+8/pw6R4QQQlj0Ys+un7d69PQVzw4ePXdVo9ZtPn/h87B66tXJUwuF\nUPniYUMuTB/9eP/p57z96KmFQghh+9JPBz/xwkfTVoRyx5x5zU3n1ysQNr97wykv1vz38Etr\n7lzF9MfOuHr15R/ec1Leny1u7X+vPuOx6SHE5CmQWLn+SZfe0L1R0ZgQPr+/0/37nLFC2L50\n7HOPDvlwxqpcJas3Pevaa06pcsDnGh5w2OZxd51318KzBw+6sFJMCCF1wr3n3z7vjEGDu1XJ\nFcLW70c++cjLY+auzV2qfttLbrisRenYXVMt/vDR+54bvSgrsVbzC84t9mqfz9v95+nOJbI9\nSwgT7mp9b/ytD5Yb//S7Xy9MK1Lv9OtvvrB+gZ2zZa358vlHBo78esm2gpWbnd27d+cj84cQ\nwtIXe17w+UnDnz5nV+Mdd1frf8X/c2SfxiFkrvx88KPPfTh9ZUbJ2kldrruyVdkc/4UAAAAA\nAADwd5Jj/inSZdCELr9x0mXTxv3fMyd/0fra/hddEqY9f8tpx895c/bzpxQKoWSrq/pXTQsh\nc+uPc0Y/eW2jUUumfXZDjZgQ0hZ+Ne6FVzt8eUGfPld0LlmxTOkW69/+cHDZi/pfVS+EIyoX\nD2HOD5PGTW2xPYQQMuY917Hp1SlNb7j54tOLbZj2fxdfmvnRq12ylk8b90WeTT+tYsPcieOW\nnJG57+LyN76of/8NIWSlb/hh4vN3N28659Pp/2yct8ZpN5328zNmzh9yWtN+G8/pd9WV1eOW\njXvskqPeu2XK6Kuq7ZNIsxuWP+niMxPq9+pSu8m4G2ts/vCaLv3ndh57dpVcIYQlL5zVqNec\ndv3vuKLmti+euaVV42/e/e65NgVCSPvkqmYd3qnd577Lm8TNfvWKtu8sWlev7rZfOktMWDNz\n3Icjp57TrnefCy89cdH/3XHpCbPyfP/WucVC2Pb5Tc2TXyh1xb3XnFZo0X/v7db4g6VTxvWu\nERO2LPxq3JRaW3/6CqtnjpuQf20IIXz3QJtWz1S6+e5eZxdZnzL2gdNvyjv54ea/8a8PAAAA\nAADw9xRzSHb7/Q13EIYQQsaaGR+P/vL7FRvS9+S2+KM6X5ZcLttDtpXo8fKbdx6XO4TTTqmy\nvtoJtz3Z95RbaoQjKjdOrrxzRKt27YsuKHbDwMk3PNJk5zuZTe9859meu3a1Va1VLMSVa5ic\nvH+f2vx/N98wpsZ9KSOurxRCCB3P6JKaGhvC5l/zVULIW7Zhctmdv7Zs06HSqgodnxh134ud\nitVq9vMzpr1563Vfn/zy/Mc7JIQQwsknFluUeOmAMVc92+pn06W9le2wst2fHzSq/rnn39vg\nhlndXyrSb9J9zeJDCBnj7u37XtX+M1/tWzOE0KlD5XWVW/3j8evb3FJ70XP9B6d3e3/EfSfG\nhRA61d1ar/6dOZ4lhJBZodeLr9/eMFcIoW3ClP92ePOT7eeenWfJoJsfW33Wm18/csoRIYRT\nj8/zQ9Vb7xhxyaudErK9Nus/H/dtle7Db+taO4TQrmOXnqmpv+6iAgAAAAAA/E84FDHvwHNu\nnfvmffe+MHllfM1Tr+t/+XFF9vk4feF7j/9r6NgFmYlHn37VDec3KBRCCBPubX3rRxk7B8S3\nGzCqb+ODWlDOgTB96gNJSf2m5CpXqUzBPHvWX+T8Zr8UCAu2bnvcrrljm7U/6YjHJn+9LdSI\nC1sXjh488JUJs5as3Lg1c+P8rKUN5m0NTfKFEEKo1rz5r3qu4ddjxmxqdPNZlX56I+6II37N\ncbvsWPH5S08P/WjGwhXrtmRsWbw5reS8ZSGU3e8sn3yyIb3ogFNPfCQrKysrKyts/iFzbblZ\nK0OrUr96WInT/z2ka/0ObS+Iazzg69uP2nmv04WTJq2sesrJu26EGvImd2xzxHOTv9oWan/z\n1ZRcSde02n0b03ptTkq8c+KvWEzcsccdteuGuLkrVCiTMWbF6hDKTJn0ZUbSJW13X5ryHTs2\nuPrpySmhU8NsL03h41rUnfdEr95Fr+ncNqlJreJH/KYLCwAAAAAA8PcWc0gC4YFsn9DnxJ7T\nznrk+lPWvtqvfdeElHe6ldz783ev7zy0SM9zr2yf+uVTfU7oGbfkP2cWCGHNrHE7mn1wz0mx\nIYTY0tUO9uQ5BsKsDx65Z+VZIxY/c0rJ3/I0ugIFCvzsRVbK+o0hZP3fhY27prS+ufeFncoV\nLRA3e+BZV/64ZWsIOwNh/vz5f9XcqampoVChQr9hNXvZPOa649u/V69Pn/NPrlS8YL4Vr1w2\nafKWLQcYuGnTplCta59bO+y9qAJV9z1tDsPyFS9VKFfW4uJ1j/rpYX4bNmz4+eUpWLBg5g8b\nNoX0jRu3JJQosOfZl3tG/fJZcufZU25jYmJCZmZmCOkbNqTlK1wgz96nCRs2bMj+0oQQ6t4y\nfnL1Zwb/35OX33fhoiNOuOKpIfe3T/zFIwAAAAAAAP53/FkbCNNHPTukyPVTH+xWLYS2sV+V\nun/oom59Kuw1oOWAcR0SEkIIofmm/xvw1IpNIezsRsVrJyUn/5ZqdwA5Hr5t/fqs5Mt+Wx0M\nIayYM2dTaL5znRtSUlbmr1y5REh//bW3QvfRr9zSIoQQwualj6zLdoJcuXJl91HVqlXD+Bkz\nMsPxPxsSn5AQs2VP69uxatXaAxw8+T+vLev01Jy7z8odQggZo15ck80Zq1SpEj7alK9ZcnKe\nfef41cNSJ/zjgge2dH/gqs9v7XH1qd8OPbXIriN+mDNne6i884j1s2evKFClSvGQt1KlxA1T\n564O7UqEEELYOm/ekt+2mL3lrVKlbNoXc5bu3hyZOXv2nFClQ5UQQkJCQtjrSq1dtSoj7C6P\nhRuc3ffxs/uGzHWTb2113IX3tlvzZPKvPSUAAAAAAMDf3J9UCJempMQ0PGfnFsD4hg1rzEmZ\nG8LegTAhISHl2Qsue3He8h82HvPEf3uW2f3BN0+e2f7VPIn12/e6vkfjEtn2tF+U41H5Gh5d\neeqU2Vm/cd6MDx+967P1IYSwfuKdD79fpmu3liHkLlAgfmPKzCWZIYTMVZ/cdP+7GdlOkLdk\nyUJrly3bdoCPal546QkrBvW9/6v1IYQQtqS8+uK4jSHE1a1bbdXH70zaEkIIqdP+9djI7Qc4\nuECBAulzZ87dHkIIG768+85XfqqI+5yxVrfLW24cclO/0St2LjLzx6+ff3zEkn3n+4VhGz+4\n9sKBuXq/9PhND79yR8nXel72+soQQih0eo/Ts4bf+9CMtBBCWDP6toc/LtejR6sQQtMuXaqO\nf+rO0T+GEELqt//615tpv20xP3Nctx61Jz9528jlmSGEbXOfvnPI2lY9zqkQQkisW7fY7Pff\nmbc9hJC1dvx9Az/ddciSkU8MmrQ6I4QQchUqXeqIXDGH31MzAQAAAAAADo2Y2OLtHtr7J6Fa\n2xBifutPQrW2e0+Sr2yT/U+VmpoaHx+/60VCQkLq5s37jSlz4jV33H7LTV3LTrj7oTE7s1jz\nWz4Zev91119+arV597c+6Z/TMw/ui2a3MXD7osmfzNzZzuqcmtDrlDO29jm3ScVCe/av5a3Y\nuFXtotnOm9iy/rcXVK9WpFzWgpnrql/3n7taxoUQ2vZ9qO2JVx9Z5dlaBVfOXVOjVfOSM7Od\nIenCXpVOuaRWoycq5q928bAhF+71UZWrhr+x+NyeLSoMrFaj6MYlmc3++Z/OIYT61z5y9Yhz\nk6uMqldmy9KtDVs32fdhjiGEcGzvf3V7q2ujSm/WKblp3tIiJ7aoNmlVNme8cviI1d0u7ljl\npfK1y+dZPX9t4hl3PnPOfhNWzmbYmreu6D6s0C1f3HdcfAj1+73yzw+O7tVj6PHvdStX9Lwn\nhn1+WvdjKwypXS79+znpTe567a4W+UIIuY+9+fnbJp7Rvuq7tavkXbaq9ElJ5T/fnDv3L53l\nF8Qe1felh78985waFavWKLBy1rKS5wx++5KyIYQQ06bf46edeHHDqq/WLrZueUzr5o1yLQ4h\nhFCkfP5JvRr+Y2nRamXCklnLEy/7d78TfvEcAAAAAAAA/zOyMtZ80GffN3/7bqq07z9K+/6j\nn17mKVYjodpJ+4xJTExcO235thDiQgjLli0rnbj/Q98KVGnSskpoedLR60ZXGvrJM+3OjA3F\naycl1w4hhJNOLja36B3vzr+l3sE8iDAmK+uAmwN/fLp1qWvG/cKBxa/8ZMVjSQf+7JkTYx5s\n9OW8e2osnpOybEdi3Xrljthz6batmfvd3DUxJWrUqZZ3yRdfb6rYvEFi7pC28KvJP5ZqenT5\nuL3m2bFh0aw5i9amxlVqfGzFzB8mfbWhfLOjyuyOlNvXzZ+RsiauYp1aiUf8tBFyx8bFs+es\nylOuVrXSO+ZP/GZb1RPqlsoVNv/82O3rf5iZsmJ7oSp1axZa/c0XK4s1aVQxfv8zJoQQQmbq\nslnfLUzLX6FWjbIFsr/P6n7Dti+dOnFerhrN6yfG7hyRtWbm+BlpFRs32jlv2LZy9rffr4st\nXat+lSI/mzdt2YzpS7JK16hZbtKVxTquG7T5jc55sztLCGHNzHGzYuu1qLkr1qYt+HLyusSm\nDcvtupAZGxdOn7k0vUClukeWSdj7H3D6uvmz5m08onKtqsXkCyOUAAAgAElEQVQ3zRqXEluv\n+a4pdmxcPGv2iqxilWtULZ4v2y/8Py1t8sC/egkAwC/JSN//f9QBAIeXHamr/+olAAA5K9J2\nwF+9hMNLVkb66rcv/cOnzVOsRpGkm/d9d94DRx8z5qqZ711Udutn1x19xqZHFw1qt1cmSxnx\n4oYTLmxcOISMFa+fW//GyiMXDWi81+GpU25u3nT6tave6VboIJaUXSD8XXYFwvsb/eEz/69b\nNfq1qbXPbJOYO4RN397Xsdm9Rzy+6N0exf7qZUWNQAgAhzmBEAAOfwIhAPwtCIT7yMpIXz3i\nsj982jzFqhc5Yb9AGDZ8fGPyWYM3Viy7Zf7GJk+Ne/PCyjFhw4hrO407aeRDHfL/OPr2s3s+\nv6JY2VxLZ6+u3GvoOw+0LRrCuLta3zE6I+zYvHjm3FytHnj71V5H5jnAGXOU/Z64XbZ+N+K1\ndY26NS+T45v8fgUKLn20ebleBcsXWDt39pY6Vwwf2E0dBAAAAAAA/p+9O4/zuU4cOP7+jmOM\nXIMwg8i6b3KkXOUYZ1bSdigdlNpqu3+di7b6WdVutal+bW1pu++bSiqbWqWDDqGIYoiQmzEz\nvz8GKYzCGPV+Ph/zx3w/38+8P+/Ph/9ej/f7w77yizcU3d0xy3a9cercP06b8W1yrWaNKuft\n6liy1SkjalQuEUKocOQ1r804b9Ync9en1qpXq8LmtYUNj/nziI65iWKlK9WsU7dqmSK7PaFd\nrSBcelvn+p+PWHpb558cbDBz5JJ/7GSL0b1hznOj7t941IhjGhbcJfKXPePJ0ZPThg05bEdv\nMtxi6aQ7b5vV9MIhh5XZZ/MqRNkzHh81ufq5Qw79rd+tFYQAsJ+zghAA9n9WEALAr4IVhD+R\nm71xyfNn7fVhi1Wok9rhsr0+7J5I2vUp28v9+usF5cuX35vzWPH23SPunLTshwNznhs18onP\n9uYlfpl5d5555sTEwfnVwRDC0kl3jrz77ZX7ZkqFrkjNMl+NOHnk2xsLeyIAAAAAAAAFI1Ew\nP/uXfLYYXfHokM7/OzV76RcrNgxp/laprcez1y6a82XKkEn19uY8Vrx998gH+gwa1nFLdqx1\n1GXDNxba8sGNr1x37awB9w9OK6wJ7J9SMi4+K7nZ8H//z6unVyrsuQAAAAAAAOxtiQKIeftd\nHsw3ECbX6zZoUP21k+/465Jug37/uy2Hk4qXqVT7sN49Ghff6V/OeWH0/VPXhpCUUqF63UMz\n+rZO33yZ+eNv/Ney7pd2Wvfyq//9ckWpukce3bdpaiJs/Oihm1/5JiyaeOuITeVDqNXn0pNb\n/XjETYvfffKJ1+fnptU/LOPQtU+P+bz5hVs2usxZ/sm4ZyZ88m1SpTqHZvRtk573KsbPn7jm\nqeInnN944YsT35+3NrVJj4EZdQ/YPNi6r958Zvz7X68rW/uwjD5tq+3gPlY/c+eDieNe7vKj\naV98+MrxE6bOWV6qUe9jezfYdpfNnCXTXhj/1udLitZs369/mypF8n8I+d/O+vmTnnr+na9W\nlazWqtfRXX+3pcvmrpr5ytMTPl6cXalBx549mx+4ZU/ZrEVTn3329dnLilZu2r1/r0apiW0m\nfGG77ye88f4XS0vW73ls38bl1s557akXp2aGqoce9YeONX54Y+WqL157dvzUb7LK1zqkW9+O\nNVPyfUr1Tjyx8dX/d9+Xp1+69T8EAAAAAADAb8U+ewVhocpni9GU5n+4+OKL/+d/x9xz0/CL\nf3DheWcM6tU49efsTZq19LPnhmc0PHTEOxvyDswff+PIywcd3uWCJz74ZuHUu4a0bnHW+J28\nQGabLUZzZ9/eo1H7ix+eseDLyWNOaNv5jKtH3v3fvG0917x7bad6na96buaS7+a8esNRjVtd\n9nreF58/cc2Ii45u13fES58u/PrNvx3b7NCr3ssKIYSVr5/fvOHxYybPX7Zw6qMX9Tj86smb\ntrt29qTxE7IP79Bmyz3OH3/jyMuOb9vx3Ec+mD9/ym0nNTvkjBd/2A31q/v6djjtvilzF336\nyAUdGv/hoUW7eAj53M5344c1r9dv1IR5Sxe8desJDRsf/8SCEEIIi586vkHLs+5/f9F38ybd\nc0bnrjfnPZi174w4tE6nK5+e/e2ij/51Zsv6Pe6YlbN1wpf2b9P1imemLfhm8s0ntGx76hWn\ntO1x7Wtzls4bP7JH8/73frN5cjPu6t2g9TkPffTtsq/fvuPk5k1PfWZx/k+pZqdO1aeOG+81\nAgAAAAAAwG9R7FuM5kmu3+O4vN9y1q9atT47d/PxpORSZVJ28te1+lw6os/m368bekWztlfd\nf/ZrQzdvSvld+tmfjxtaLRHC+u4lqve+5cnRPQY3P+H87rfe8u2R5424qvZ2o6148uorJje7\nddarw6onhTDimUF1+3+2efHglJEnjip69Sfvn1szKYSQe97NHRucNfqEz69tGkIIWSubjJj5\nYP9yIYQrWvdJP+P2V/5yb+/EpLF3f3fCE5/d3aNICCGs+XRa5vb/JvM++WTVQV3r/Ghp4deJ\nvu/+d1Tr5BDWn3J5izZ/Gj00Y1TroiGE8F3ycROnXN6waAi5xx1Yr/Xfx351wv/U3PlD2Pnt\nZL31l7PvKnLue++OPqRECFlnX9OuyTlXDep5b++ccf96LOXcDyaMah5CCLnLpn28JoQQZv7t\n7GsXD3zus3/1KhNCziWHZ9S78KJ/H/384Mp5j63eJTNfOr1qCOGcGu1qXnx7owdmvX5ipRDC\nKZWaNL7lvi9Ovap2+PLvg89dePqUaSObFw8hhCu6/L72GcNPPOrOrvk8pbp16+beN/2TEI7Y\n8T8+AAAAAADAr1cUSwh3GQhDCEv+c9MFF938wrQF32/M3Xqwwh9fX3pb5539yaZv33/ppSmz\nFy5dtTHnq2K506fPCGFzIGz1+/7V8h5DiebN6+VM/urrEBrld/kpL7+8ststp1TPW89Xtt+p\nR6c++GEIIYSPnnziiwr1v37g+mtzc3Nzc3NDZlKpmVPeWxGalgshFD2if99yeUOkNm9eY93k\nr74NoXLVatW+f+rRh99tOqBNeko4oFGz7YtkWL58eShTpsyPjjU+/tTWyXnTbnX6iU1HPT5+\n9qjWDUIIoV7f/g3zHmOiefOm4Z6v5oVQc+cPYee38/m4cXNbnnrGISVCCCEUazDstI7Drxn3\nYejdrlq18gsn/vvpaVV7NzuweKJ8s6blQwiZr4z7qMpxd/TKm2dStdOG9j73pHH/yRl8TFII\nIbQ6ql/VvKnXaNTogFD2qH6b/wEaNGqUdPO8eSHUnv/sE++VKtv6hdHXPp+bG3Jzc7/fUHbJ\nlClzQ9d8nlKZMmXCiuXL8/sXAwAAAAAA+HVKRBEId71T6NJHz/79377rd9O957QoO3DMGy/c\nf+0fGpZOG3D9eYfs7C9WTrywxcHdr3jyvXnLN+SEkJSUtHyboFSq1Jb36oWiRYuGrKys/K+/\nITNzRdnKlUts+ZyoUqXy5l8zMzNDieRimzZtys7OzsnJyal8xLnDBzbY/G1yqVJb8+cPV2px\n+VP/6rX85qPrlq9Y/4gTr35sxprtL5mamhpWrlz5o2NVqlTZ+ntaWlpYuHBhvje0s4eQz+0s\nXLgwpKWlbb1MpfT0oosXLNgUinT935duaPXp8IwaqVWaZpw26qWvNu7g9KLp6QduXLBgy+af\nP5nWDx+TihZN5E0y7/kVz3t+2Tk5OaXbnjp8cOuUfJ/SypUrQ7nU1O2fGgAAAAAAwK9cFBuM\n/owVhLn/ee6lln+ed+Wx4bbbk9bV79T7yE69M6r3qv/Q25vOqLvDv1j19A03rxjy5pe3dCge\nQggr//X2yEd2OY/ETntsclpaue8XL14fwuaotmjR4s1fVapUKSQaDRwxovkux//BAQ1PuvGZ\nk27ctGz2fx4bMez4LsvSvhnT4ced9KBGjUrPmzlzQ2iavPXYokWLQiib93tmZmZIPyw9v6vs\n9CHkczvp6enh48zMEDYv11uSmbmpcoOqRUMIqW3Ovn382bdv+PbT18ZeNqRv76zpn16dnp4e\nFmVmhlAj7/RNmZlLileteuDPfxSVKlUK6w7u/ecRXbfrxDt/SrNmzUo06df4518FAAAAAADg\n16IgVhAWyKrEPbLLFYTLFi5MrlOnYgily5RZtXx5TgghVOrU6XeffDJzJ3+xfu3a3GIlSuSl\nx6wZd9zz5s+YR9myZcOqVat29FXbjIwyE+6/f0FOCCGElc+NfXrLcsQWRw84eOY/r39icc6W\nc9fMfGf6svyus+Hz/36wLCeEULR8nSPOOO3I1EVz56796UlFO/boUmTypP/mbHPsk4fvfW9D\n3v1NvefB6bV6ZNTJ7zo7fwg7v536PXse/MEDd3+Yd5msmXfdO6lyz54tQlj28Tufrw4hhORK\njXqdO6ht0ldz5+WGtO49m2c+evf4vIeWs3Ds3S8W696zw67XhG51cL8BLVY++Nfb525dxLlh\n/tsfLMj3Kc2bNGn+IT17/IIMCQAAAAAA8KsRwwLCn7GCMDc3KSkphJBcv371y555fvmAfmW+\nmfDGzIqdKu7kLw78/dDfDx9yTNfvjj0kec6Ex95blVZs1/Mo1/HIFsNuOvf0cGT1lNp9Lj25\n1TZfDfjL9Xe2O6dd53eOblV85riX55erEPKmVPTwax657sO+JzRp88jRnQ4utmTG2+/MbzX6\ntXZN87mdxS+e2WdIuTYdmx9UYsnkxx4rd/pjvUttd1bp/mcN+tOgB179W6eMLXOvnvvCMYd+\n0feIKosnPvTsxlOfvLRNvo9u5w9h57dTrP3Vt58xrn+XtrOO71Ltu0mPPr3kmAef7n1ACGvn\nPjCw/9k1DuvQKD3Mm/DIq43Oe7VTIoR6F95+1dPdB7ReOKhX3Q3vPfnwrMNv/s9JlfOb1U/V\nvfiBO6b0/GOLFuOP6d7ggOVfvjv5s/TLXzis5c6f0qwHH5ze5bwXd/DiRgAAAAAAgF+/KN5B\nWGTEiBH5nrBx9sRxmw4f0rFqqFq7zITLTjr/xn/ecuPDmd1G3XF2s9I7/pPSTQcOzqieWLMu\nqfIhg67/x7mtSlao36FTg7ygWLb2YZ2bVN66zK1YevPObQ8uFUKFw/5wdMPktavWbthUru7h\nzdKLhZBar33nhgeGkKjQ+oSTu1dNrM4p16Dnn/46YN0/x649auSJzYqHULxqh8HDjmtVIWdd\nVvGqh/Q597obhrTa0vsqNuzUoV75rbMqWbN15xbVDqh55NDBXdITa1dnlaqdcd7Nt5zTusyO\nHsvvmqQ8f8E94eSTW5YOYf74G+8t9pdZTxxfcu2a7Artz/zbPy5pv/UlfCVrtOrcsvoPe5Ee\n2LBz+3qp+TyEfG6nZO0+Q05qVyFr1cbkWt3OveUfl7avGEIIB9Trfebxh5XPXbU6p3zj/peN\nueHk+iVCCKFY9c6nndq9ali5rkhah9NHjxnZu9rWB/uT55xat33nRj8s+6vYoFOH+uVDKHLg\nIccPG9whPWn9hqRKTboOveamCzpWCqHoTp7S+levOPnZw8fcdfRBRfL9T/Nrl7VgamFPAQDI\nT272xsKeAgCwCzlZ223ZBADsf1JqdyvsKexncrPXffHyXl9AWKRk+RLV2xX2vf1IIjc39+ef\nvenb6a+9MX1F6aZdMppW/AV7We6hlV/MWlurbpWkEEJYM2FYo+5Ths768MoCXsWWPePJ0W9V\nGTb08NTw1vlVOnxx47oXBpXYKyMXyu3sDdkzHh81ufo5Qw4tW9gzKWBr372jsKcAAOQne+Pq\nwp4CALALm9YsKewpAAC7lpoxurCnsH/Jzd64bNxFe33YouV/V/aw8/f6sHtiV1uMrpo9cdyb\nny0tflCbbr1apRWt1DTj2Hw28Cwg2dOv7TpweZvO9Usve//5xz868Mqnzin4nFakwYDLGxTI\nyIVyO3tDkQYDryyYRwIAAAAAALA/iGKH0fwD4dLnT2sz8N65G0IIIZHa/obX37ioWWHsLZl6\n9Ni36r/ywsSPF9U99Zbzu/dqUXnfzuKgHhcPX9Z0l69r/LkK+3YAAAAAAADYsf2v5hWA/LYY\n/Xh4ozYvZjz9yMgjqq6bdu8Z/S4pcuu3Tw48YF9ODwqBLUYBYD9ni1EA2P/ZYhQAfhVsMfoT\nudkbl42/ZK8PW7R8rbLt/rTXh90T+bxIMGv69C/7XTaqR+3SySmV2px9/dAaH3/85b6bGQAA\nAAAAAOxTicTe/9n/ViXmEwi/X768VMWKxbd8rFSp0vLly/fFnAAAAAAAAIACkv+L9bIWfPTG\nG5t//372io3ffPjGG5t3JC1WtfnhdcoV6NwAAAAAAABg30kk9v5qv4IYcw/lHwhXPnPBEc9s\ne+CHjxX++PrS2zoX0KwAAAAAAABg39vvYl5ByCcQlh/8yNw+63f2bZEyVQpiPgAAAAAAAFBY\nCiIQ7njM9bOfuv66+95dnFLvqAtGnHVo6k++3jjvpVtvGPvGVzlpLfufc9EJzcqGEELI/uaV\nm665c+L8UKPrH4df2CU9n7cJ5iOfQJhUukrN0rs1KAAAAAAAAPz67LM+mPXWJV2HTBv49wv7\nLHv48p4nl5z5/OBK237/4oXHjE0dctwfe655b8wlHYckf/P4gNIhfDqq1zGPtxv95zMSL18z\noF/ivfcuqrM7M8p/i1EAAAAAAACIRGKfFcKN4+66N/XCj24cXDuEjCJTK48aO3/wJQdtc8IR\no9/sXbJkCCG0X/Xk6DGLVoVQOrx9z/8tP+2hO4e1T4T+B3520Cl3v3fRX1vvxoR2b90hAAAA\nAAAA/NYkQqIgfra/0IKZMxMtWtQOIYSQ0qJF3VmzZv/4hJIlS86868TOHdrWazCy5D/+OSQ9\nhLB+5syvm7dokQghhESLFs3mzpqVtVu3aQUhAAAAAAAAhJBIKtvp8m0PrJ/31oZ5b/3SYZJr\ntC9Ro/3Wj1lLPtv+nDVr1qSkpGz+ULJkyTWrV293TnrX84YfvHjO67eO/MtNr/f/Z49yq9es\nCSkpJfK+TSlZMnf16rUhlP2l8xMIAQAAAAAAIIQQcnO+n/TX7Y7+4k1HN8ybvGHe5K0fi6bW\nTD6o/U/OSUtLWzYtc0MIySGEhQsXVklL226c0rXaHlErHNGt5fKJNce+dmePARXS0opnZi4K\noWoIIXPhwpS0tN2og8EWowAAAAAAALBZogB+dqRCl65NJ/77wQU5IaydPPaJVd27N/vR9zOf\n+/e7K0IIIWQvenvKnPTq1YuEkOjY9cgPHxo7MyuErJljH/64e/efdsefyQpCAAAAAAAAyPOL\n1wvu5pi1h40e+sjAxnVurbpuzsq2Y97MSA4hfP/c+f3e7PbCTb1LVSw1++yW1RZVqJq04PMl\nBw8b+39tQgjhwBOvvfK+Hm1qP1oz96tFtYdPOLb8bk4oNzd3928HfovWvntHYU8BAMhP9sbt\n9+QHAPYvm9YsKewpAAC7lpoxurCnsH/Jzc5aMfHPe33YouVqlm595o6+yV4xd9qMb5NrNWtU\nOe/FglkLP5q8pHL7ZmlFQwhhw9JZn8xdn1qrXq0KydvMcs3X06cvCNWbNK12wO7WTCsIAQAA\nAAAAIIQQEgWwgnDnYxYpd3DLdgdvc6BYevPO6Vs/JVese0jF7Yc7oHqzdtX3bEreQQgAAAAA\nAAARsYIQAAAAAAAAQgghJArgHYQF8VrDPSMQAgAAAAAAQJ6CqHn7XSEUCAEAAAAAACCPQAgA\nAAAAAADxiKIPCoQAAAAAAAAQQgghUQA1ryDG3EMCIQAAAAAAAIQQQkgUQMzb7/qgQAgAAAAA\nAACbRbHHqEAIAAAAAAAAeQRCAAAAAAAAiEdBbDEqEAIAAAAAAMD+ab9LeQVDIAQAAAAAAIAQ\nQsGsINz/qqNACAAAAAAAAHlsMQoAAAAAAAAREQgBAAAAAAAgEolI+qBACAAAAAAAAJtFUQgF\nQgAAAAAAAAghhJAogJi33/VBgRAAAAAAAABCCCEkCqDmFcSYe0ggBAAAAAAAgBBCwawgFAgB\nAAAAAACA9bOfuv66+95dnFLvqAtGnHVo6o+/zV32/n03jXlu2vLSDboNu/jMwyoXCSGEt67r\nctWr2XlnpPQYPe6yNrt1aYEQAAAAAAAA8uyrFYRZb13Sdci0gX+/sM+yhy/veXLJmc8PrrTN\n1x+P6HfONz2HnjyszJwHruze/bsp065uFEJYOuPNTYe/fG23IiGEIlVq7+6EBEIAAAAAAAAI\nIRTMFqM7GnPjuLvuTb3woxsH1w4ho8jUyqPGzh98yUE/fP+7S8ZNLlUqKYQQOq+ZUPrBKStD\nozIhhBAqNujUufMeFr6kPftzAAAAAAAA+M1IFMDPDiyYOTPRokXeEsCUFi3qzpo1+0ffl9xc\nB0NYMfG1aU3atyuz5ZsPbxvQs/cxp11+z7tLcnb3Jq0gBAAAAAAAgJBIJB3Qaui2R7IWvp+1\n8INfOk6x9JbF0g/Z+jF72Zfbn7NmzZqUlJTNH0qWLLlm9eodDbVu+t+Pvmj+pU/9s0He5/ZX\nvjZ2cW7W6vnvjf1Ll26L3v7gyia7sxpQIAQAAAAAAIAQcnPWvn/3Tw/+8k1HszI/zMr8cOvH\nImWqFktv+ZNz0tLSlk3L3BBCcghh4cKFVdLSthtn1dTRfY4f1/exF89vkrz5UMUGnTo3CCGE\nbr0qzC4//MU5VzbZnRcR2mIUAAAAAAAAQsHsL7rjXUYrdOnadOK/H1yQE8LayWOfWNW9e7Mf\nn/D92yO7Hzu+32MvXtzigB3MdM1Hkz9al5Z24O7dpxWEAAAAAAAAEELYnfWCuzlm7WGjhz4y\nsHGdW6uum7Oy7Zg3M5JDCN8/d36/N7u9cFPvUpOvGjDi4xKtky/o9VwIIdQ+48G7T6ga3rym\ny/CJ2WHT6q8/m5105C3PnlB292YkEAIAAAAAAMA+VrbrjVPn/nHajG+TazVrVLlECCGEkq1O\nGVGjcokQQsNzHn99wKatJ5eqVT6EEBoe8+cRHXMTxUpXqlmnbtUyRXb32onc3Nw9nD78xqx9\n947CngIAkJ/sjTt8aTcAsB/ZtGZJYU8BANi11IzRhT2F/UzOptX/vW2vj1qkTHpK42P3+rB7\nwgpCAAAAAAAACCEUzBajO3oHYeESCAEAAAAAACCEEBIFEPMKYsw9JBACAAAAAABACKFgVhAW\nyKrEPZJU2BMAAAAAAAAA9h0rCAEAAAAAACDPfrfaryAIhAAAAAAAABBCiGWLUYEQAAAAAAAA\n8hREzBMIAQAAAAAAYL9UIIv99rs+KBACAAAAAABACCGRVKx6u70+alLxUnt9zD0kEAIAAAAA\nAEAIiaRiaS0KexL7QlJhTwAAAAAAAADYdwRCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAi\nAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAA\nAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQ\nAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAA\nREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAA\nAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIi\nEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAA\nAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgB\nAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABA\nRARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAA\nAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICIC\nIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAA\nAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAA\nAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABE\nRCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAA\nAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQ\nAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAA\ngIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEA\nAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBE\nBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAA\nAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIh\nAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAA\niIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABERAz6Qp8AAA2ZSURBVCAEAAAAAACAiAiE\nAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAA\nICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAA\nAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABAR\ngRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAA\nAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAI\nAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAA\nIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAA\nAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABER\nCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAA\nAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQA\nAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAg\nIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAA\nAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGB\nEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAA\nAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgA\nAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAi\nIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAA\nAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREI\nAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAA\nQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAA\nAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAi\nAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAA\nAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQ\nAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAA\nREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAA\nAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIi\nEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAA\nAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgB\nAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABA\nRARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAA\nAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICIC\nIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQAgAAAAAAQEQEQgAAAAAA\nAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAABARgRAA\nAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEAAAAAACAiAiEAAAAAAABE\nRCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBEBEIAAAAAAACIiEAIAAAA\nAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAAAAAQEYEQAAAAAAAAIiIQ\nAgAAAAAAQEQEQgAAAAAAAIiIQAgAAAAAAAAREQgBAAAAAAAgIgIhAAAAAAAAREQgBAAAAAAA\ngIgIhAAAAAAAABARgRAAAAAAAAAiIhACAAAAAABARARCAAAAAAAAiIhACAAAAAAAABERCAEA\nAAAAACAiAiEAAAAAAABERCAEAAAAAACAiAiEAAAAAAAAEBGBEAAAAAAAACIiEAIAAAAAAEBE\nBEIAAAAAAACIiEAIAAAAAAAAEREIAQAAAAAAICICIQAAAAAAAEREIAQAAAAAAICICIQAAAAA\nAAAQEYEQAAAAAAAAIiIQAgAAwP+3b8c2AIAADMPE/0eXL2CIfUH3qAAAACECIQAAAAAAAIQI\nhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACECIQAAAAAAAAQIhACAAAAAABA\niEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAAQIhACAAAAAAAACECIQAAAAAA\nAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACECIQAAAAAAAAQIhACAAAA\nAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAAQIhACAAAAAAAACECIQAA\nAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACECIQAAAAAAAAQIhAC\nAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAAQMjZ9nsDAAAAAAAA\n8IgHIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACECIQAAAAA\nAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAAQIhACAAA\nAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACECIQA\nAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAAQIhA\nCAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAAAACE\nCIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAAAAAA\nQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEAAAAA\nAACECIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQAgAA\nAAAAQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAhAiEA\nAAAAAACECIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAAECIQ\nAgAAAAAAQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAAAAAh\nAiEAAAAAAACECIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAAAAAA\nECIQAgAAAAAAQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgAAAAA\nAAAhAiEAAAAAAACECIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiEAAAA\nAAAAECIQAgAAAAAAQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECIQAgA\nAAAAAAAhAiEAAAAAAACECIQAAAAAAAAQIhACAAAAAABAiEAIAAAAAAAAIQIhAAAAAAAAhAiE\nAAAAAAAAECIQAgAAAAAAQIhACAAAAAAAACECIQAAAAAAAIQIhAAAAAAAABAiEAIAAAAAAECI\nQAgAAAAAAAAhF1u2fSyPsa96AAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 600,
       "width": 1200
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "pathways_sum_plot[[i]] +  scale_y_discrete(labels = label_wrap(25))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "bc8fa8ea-db7e-41cc-80f1-33a034235a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot factor values for all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "11cccb41-8fe5-4217-9558-a3186731b3bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_detail_plot = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "e4afadd0-7b99-4e29-8042-5d4c8c71b948",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Visualized the exact factor values of the genes\n",
    "for(i in unique(pathways_vis_genes$variable)){\n",
    "    # Specific Text Descriptions:\n",
    "    xlabel = xlab('Gene') \n",
    "    ylabel = ylab('View')\n",
    "\n",
    "    plot_data_cluster = unique(pathways_vis_genes[pathways_vis_genes$variable == i,c('gene', 'variable', 'type','view',  'variable_name', 'factor_value', 'factor_value_direction')])\n",
    "\n",
    "    pathways_detail_plot[[i]] = ggplot(plot_data_cluster, aes(gene,  view, fill= factor_value)) + \n",
    "        plot_config_heatmap + \n",
    "        geom_tile() + \n",
    "        scale_fill_gradient2(low = \"#1D2ED8\", mid = \"white\", high =\"#D8911D\", midpoint = 0)  + \n",
    "        scale_x_discrete(position = \"top\") +\n",
    "        theme(axis.text.x = element_text(angle = 90)) +\n",
    "        xlabel +\n",
    "        ylabel\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "eb692dd3-fc2f-420b-b3a1-198964530e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pathways_detail_plot[[i]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "babfbb17-7380-42ba-84c7-a7ab2ef35a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine plot and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "b1d8db71-8f27-4a31-baa0-6111ccf1e551",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_plot = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "bd4a76d8-c89a-496d-96ef-5377d60862a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in 1:length(pathways_detail_plot)){\n",
    "    combined_plot[[i]] = ggarrange(pathways_sum_plot[[i]],\n",
    "              pathways_detail_plot[[i]] + theme(axis.text.x = element_blank(), axis.title.x =element_blank()), align = 'v', ncol = 1)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "29a27873-76c0-4219-be95-0e3a52f88bbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combined_plot[[i]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "274e7e24-a024-4b5e-b5b3-49b4ed912d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d0d6a79b-f80c-461c-803e-b15ee5eb821d",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_name = paste0( \"FIG06_Pathways_and_Genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "665879d0-eba1-4902-980e-40ffa9da88d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sizes of the plot\n",
    "width_par = 8.07\n",
    "height_par = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "b067c7a7-213e-4606-9dbd-298d37670d89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/06_figures/', figure_name, '_',   mofa_name, '.pdf'), width =width_par, height =height_par)\n",
    "\n",
    "for(j in 1:length(combined_plot)){\n",
    "        print( combined_plot[[j]])\n",
    "        }\n",
    "dev.off()   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca14ac62-5f47-4fde-ad71-f36d1cbb6832",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "337a3db7-43f2-49ff-8f5f-c6f9209c84c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
